{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba51447c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.gridspec as gridspec\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from matplotlib import colors as colors_mpl\n",
    "import seaborn as sns\n",
    "import zipfile\n",
    "from CRISPResso2 import CRISPRessoShared\n",
    "from CRISPResso2 import CRISPResso2Align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ece16f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collab New Data - Compound Ref Alignment\n",
    "crispresso_output_folder = \"CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_AAVS1_PE2max_rep1/\"\n",
    "# crispresso_output_folder = \"/CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_AAVS1_PE2max_rep2/\"\n",
    "# crispresso_output_folder = \"/CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_AAVS1_PE2max_rep3/\"\n",
    "# crispresso_output_folder = \"/CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_CCR5_PE2max_rep1/\"\n",
    "# crispresso_output_folder = \"/CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_CCR5_PE2max_rep2/\"\n",
    "# crispresso_output_folder = \"/CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_CCR5_PE2max_rep3/\"\n",
    "# AAVS1 TwinPE reference sequence\n",
    "twin_seq = 'CGCCGGGAACTGCCGCTGGCCCCCCACCGCCAGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTGGGAAGTCCGCAGCTCCGAGGCGCCCAGTGGAAATCGCCAGATGAGGGCCTCCTC'\n",
    "wt_seq =   'CGCCGGGAACTGCCGCTGGCCCCCCACCGCCCCAAGGATCTCCCGGTCCCCGCCCGGCGTGCTGACGTCACGGCGCTGCCCCAGGGTGTGCTGGGCAGGTCGCGGGGAGCGCTGGGAAATGGAGTCCATTAGCAGAAGTGGCCCTTGGCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTGGGAAGTCCGCAGCTCCGAGGCGCCCAGTGGAAATCGCCAGATGAGGGCCTCCTC'\n",
    "# compound_ref_seq = 'CGCCGGGAACTGCCGCTGGCCCCCCACCGCCCCAAGGATCTCCCGGTCCCCGCCCGGCGTGCTGACGTCACGGCGCTGCCCCAGGGTGTGCTGGGCAGGTCGCGGGGAGCGCTGGGAAATGGAGTCCATTAGCAGAAGTGGCCCTTGGAGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTGGGAAGTCCGCAGCTCCGAGGCGCCCAGTGGAAATCGCCAGATGAGGGCCTCCTC'\n",
    "# CCR5 TwinPE reference sequence\n",
    "# twin_seq = 'ACAATCGATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCTTTAAAAGCCAGGACGGTCACCTTTGGGGTGGTGACAAGTGTGATCACTTGGGAGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTGCTCTCATTTTCCATACAGTCAGTATCAATTCTGGAAGAATTTCCAGACATTAAAGATAGTCATCTTGGGGCTGGT'\n",
    "# compound_ref_seq = 'ACAATCGATAGGTACCTGGCTGTCGTCCATGCTGTGTTTGCTTTAAAAGCCAGGACGGTCACCTTTGGGGTGGTGACAAGTGTGATCACTTGGGTGGTGGCTGTGTTTGCGTCTCTCCCAGGAATCATCTTTACCAGATCTCAAAAAGAAGGTCTTCATTACACCTGCAAGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTGCTCTCATTTTCCATACAGTCAGTATCAATTCTGGAAGAATTTCCAGACATTAAAGATAGTCATCTTGGGGCTGGT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a4f2c015",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Anzalone Data\n",
    "\n",
    "#### PE3 12bp ins sample\n",
    "# crispresso_output_folder = \"../04_crispresso_on_anzalone/CRISPRessoBatch_on_fig2b_hek3_pe3_ins_12bp/CRISPResso_on_Fig2b_HEK3_PE3_ins_12bp_rep1/\"\n",
    "# PE3 Reference Sequence for 12bp ins\n",
    "# Full twin pe ref\n",
    "# twin_seq = 'GCTGCCTAGAAAGGCATGGATGAGAGAAGCCTGGAGACAGGGATCCCAGGGAAACGCCCATGCAATTAGTCTATTTCTGCTGCAAGTAAGCATGCATTTGTAGGCTTGATGCTTTTTTTCTGCTTCTCCAGCCCTGGCCTGGGTCAATCCTTGGGGCCCAGACTGAGCACGGGAGTGCAGGTGTGATGGCAGAGGAAAGGAAGCCCTGCTTCCTCCAGAGGGCGTCGCAGGACAGCTTTTCCTAGACAGGGGCTAGTATGTGCAGCTCCTGCACCGGGATAC'\n",
    "# Truncated twin pe ref - matches wt ref seq length\n",
    "# twin_seq = 'GCTGCCTAGAAAGGCATGGATGAGAGAAGCCTGGAGACAGGGATCCCAGGGAAACGCCCATGCAATTAGTCTATTTCTGCTGCAAGTAAGCATGCATTTGTAGGCTTGATGCTTTTTTTCTGCTTCTCCAGCCCTGGCCTGGGTCAATCCTTGGGGCCCAGACTGAGCACGGGAGTGCAGGTGTGATGGCAGAGGAAAGGAAGCCCTGCTTCCTCCAGAGGGCGTCGCAGGACAGCTTTTCCTAGACAGGGGCTAGTATGTGCAGCTCCT'\n",
    "\n",
    "#### PE3 108bp ins sample\n",
    "# crispresso_output_folder = \"../04_crispresso_on_anzalone/CRISPRessoBatch_on_fig2b_hek3_pe3_ins_108bp/CRISPResso_on_Fig2b_HEK3_PE3_ins_108bp_rep1/\"\n",
    "# # Full PE3 Ref Seq for 108bp ins\n",
    "# twin_seq = 'GCTGCCTAGAAAGGCATGGATGAGAGAAGCCTGGAGACAGGGATCCCAGGGAAACGCCCATGCAATTAGTCTATTTCTGCTGCAAGTAAGCATGCATTTGTAGGCTTGATGCTTTTTTTCTGCTTCTCCAGCCCTGGCCTGGGTCAATCCTTGGGGCCCAGACTGAGCACGGGAGTGCAGGTGGAAACCATCTCCCCAGGAGACGGGCGCACCTTCCCCAAGCGCGGCCAGACCTGCGTGGTGCACTACACCGGGATGCTTGAAGATGGAAAGAAATTTTGATGGCAGAGGAAAGGAAGCCCTGCTTCCTCCAGAGGGCGTCGCAGGACAGCTTTTCCTAGACAGGGGCTAGTATGTGCAGCTCCTGCACCGGGATAC'\n",
    "\n",
    "#### TwinPE 108bp ins sample\n",
    "# crispresso_output_folder = \"../04_crispresso_on_anzalone/CRISPRessoBatch_on_fig2b_hek3_twinpe_ins_108bp/CRISPResso_on_Fig2b_HEK3_twinPE_ins_108bp_rep1/\"\n",
    "# TwinPE Reference Sequence for 108bp ins\n",
    "# twin_seq = 'CCCAGCCAAACTTGTCAACCAGTATCCCGGAAATTTCTTTCCATCTTCAAGCATCCCGGTGTAGTGCACCACGCAGGTCTGGCCGCGCTTGGGGAAGGTGCGCCCGTCTCCTGGGGAGATGGTTTCCACCTGCACTCCCGTGCTCAGTCTGGGCCCCAAGGATTGACCCAGGCCAGGGCTGGAGAAGCAGAAAAAAAGCATCAAGCCTACAAATGCATGCTTACTTGCAGCAGAAATAGACTAATTGCATGGGCGTTTCCCTGGGATCCCTGTCTCCAGG'\n",
    "\n",
    "#### TwinPE 38bp ins sample Fig1c A38 B38\n",
    "# crispresso_output_folder = \"../04_crispresso_on_anzalone/CRISPRessoBatch_on_fig1c_twinpe_38bp_ins/CRISPResso_on_Fig1c_attB_A38_B38_twinPE_ins_38bp_rep1/\"\n",
    "# twin_seq = 'AAGGCATGGATGAGAGAAGCCTGGAGACAGGGATCCCAGGGAAACGCCCATGCAATTAGTCTATTTCTGCTGCAAGTAAGCATGCATTTGTAGGCTTGATGCTTTTTTTCTGCTTCTCCAGCCCTGGCCTGGGTCAATCCTTGGGGCCCAGACTGAGCACGGGCTTGTCGACGACGGCGGTCTCCGTCGTCAGGATCATCCGGGATACTGGTTGACAAGTTTGGCTGGG' # full length twinPE ref - shorter than WT Ref due to 90bp deletion but actually contains more of the original bases\n",
    "# twin_seq = 'AAGGCATGGATGAGAGAAGCCTGGAGACAGGGATCCCAGGGAAACGCCCATGCAATTAGTCTATTTCTGCTGCAAGTAAGCATGCATTTGTAGGCTTGATGCTTTTTTTCTGCTTCTCCAGCCCTGGCCTGGGTCAATCCTTGGGGCCCAGACTGAGCACGGGCTTGTCGACGACGGCGGTCTCCGTCGTCAGGATCATCCGGGATAC' # Issues with finding Perfect PE so can't use - truncated twinPE ref to base match WT Ref - might help with alignment and visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10d66097",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_refpos_values(ref_aln_seq, read_aln_seq):\n",
    "    \"\"\"\n",
    "    Given a reference alignment this returns a dictionary such that refpos_dict[ind] is the value of the read at the position corresponding to the ind'th base in the reference\n",
    "    Any additional bases in the read (gaps in the ref) are assigned to the first position of the ref (i.e. refpos_dict[0])\n",
    "    For other additional bases in the ref (gaps in the read), the value is appended to the last position of the ref that had a non-gap base (to the left)\n",
    "    For example:\n",
    "    ref_seq =  '--A-TGC-'\n",
    "    read_seq = 'GGAGTCGA'\n",
    "    get_refpos_values(ref_seq, read_seq)\n",
    "    {0: 'GGAG', 1: 'T', 2: 'C', 3: 'GA'}\n",
    "    Args:\n",
    "    - ref_aln_seq: str, reference alignment sequence\n",
    "    - read_aln_seq: str, read alignment sequence\n",
    "    Returns:\n",
    "    - refpos_dict: dict, dictionary such that refpos_dict[ind] is the value of the read at the position corresponding to the ind'th base in the reference\n",
    "    \"\"\"\n",
    "    refpos_dict = defaultdict(str)\n",
    "\n",
    "    # First, if there are insertions in read, add those to the first position in ref\n",
    "    if ref_aln_seq[0] == '-':\n",
    "        aln_index = 0\n",
    "        read_start_bases = \"\"\n",
    "        while aln_index < len(ref_aln_seq) and ref_aln_seq[aln_index] == '-':\n",
    "            read_start_bases += read_aln_seq[aln_index]\n",
    "            aln_index += 1\n",
    "        refpos_dict[0] = read_start_bases\n",
    "        ref_aln_seq = ref_aln_seq[aln_index:]\n",
    "        read_aln_seq = read_aln_seq[aln_index:]\n",
    "        \n",
    "    ref_pos = 0\n",
    "    last_nongap_ref_pos = 0\n",
    "    for ind in range(len(ref_aln_seq)):\n",
    "        ref_base = ref_aln_seq[ind]\n",
    "        read_base = read_aln_seq[ind]\n",
    "        if ref_base == '-':\n",
    "            refpos_dict[last_nongap_ref_pos] += read_base\n",
    "        else:\n",
    "            refpos_dict[ref_pos] += read_base\n",
    "            last_nongap_ref_pos = ref_pos\n",
    "            ref_pos += 1\n",
    "    return refpos_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "44627f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(twin_aln_seq)\n",
    "# print(comp_aln_seq_pe)\n",
    "# print(comp_aln_seq_wt)\n",
    "# print(wt_aln_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d4127c95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing CRISPRessoBatch_on_aavs1_ccr5_compound_aln/CRISPResso_on_eeBxbI_50attP_AAVS1_PE2max_rep1/\n",
      "Found 167 base changes (via compound alignment): [(31, 'C', '-'), (32, 'C', '-'), (33, 'A', '-'), (34, 'A', '-'), (35, 'G', '-'), (36, 'G', '-'), (37, 'A', '-'), (38, 'T', '-'), (39, 'C', '-'), (40, 'T', '-'), (41, 'C', '-'), (42, 'C', '-'), (43, 'C', '-'), (44, 'G', '-'), (45, 'G', '-'), (46, 'T', '-'), (47, 'C', '-'), (48, 'C', '-'), (49, 'C', '-'), (50, 'C', '-'), (51, 'G', '-'), (52, 'C', '-'), (53, 'C', '-'), (54, 'C', '-'), (55, 'G', '-'), (56, 'G', '-'), (57, 'C', '-'), (58, 'G', '-'), (59, 'T', '-'), (60, 'G', '-'), (61, 'C', '-'), (62, 'T', '-'), (63, 'G', '-'), (64, 'A', '-'), (65, 'C', '-'), (66, 'G', '-'), (67, 'T', '-'), (68, 'C', '-'), (69, 'A', '-'), (70, 'C', '-'), (71, 'G', '-'), (72, 'G', '-'), (73, 'C', '-'), (74, 'G', '-'), (75, 'C', '-'), (76, 'T', '-'), (77, 'G', '-'), (78, 'C', '-'), (79, 'C', '-'), (80, 'C', '-'), (81, 'C', '-'), (82, 'A', '-'), (83, 'G', '-'), (84, 'G', '-'), (85, 'G', '-'), (86, 'T', '-'), (87, 'G', '-'), (88, 'T', '-'), (89, 'G', '-'), (90, 'C', '-'), (91, 'T', '-'), (92, 'G', '-'), (93, 'G', '-'), (94, 'G', '-'), (95, 'C', '-'), (96, 'A', '-'), (97, 'G', '-'), (98, 'G', '-'), (99, 'T', '-'), (100, 'C', '-'), (101, 'G', '-'), (102, 'C', '-'), (103, 'G', '-'), (104, 'G', '-'), (105, 'G', '-'), (106, 'G', '-'), (107, 'A', '-'), (108, 'G', '-'), (109, 'C', '-'), (110, 'G', '-'), (111, 'C', '-'), (112, 'T', '-'), (113, 'G', '-'), (114, 'G', '-'), (115, 'G', '-'), (116, 'A', '-'), (117, 'A', '-'), (118, 'A', '-'), (119, 'T', '-'), (120, 'G', '-'), (121, 'G', '-'), (122, 'A', '-'), (123, 'G', '-'), (124, 'T', '-'), (125, 'C', '-'), (126, 'C', '-'), (127, 'A', '-'), (128, 'T', '-'), (129, 'T', '-'), (130, 'A', '-'), (131, 'G', '-'), (132, 'C', '-'), (133, 'A', '-'), (134, 'G', '-'), (135, 'A', '-'), (136, 'A', '-'), (137, 'G', '-'), (138, 'T', '-'), (139, 'G', '-'), (140, 'G', '-'), (141, 'C', '-'), (142, 'C', '-'), (143, 'C', '-'), (144, 'T', '-'), (145, 'T', '-'), (146, 'G', '-'), (147, 'G', '-'), (148, '-', 'A'), (149, '-', 'G'), (150, '-', 'G'), (151, '-', 'T'), (152, '-', 'T'), (153, '-', 'T'), (154, '-', 'G'), (155, '-', 'T'), (156, '-', 'C'), (157, '-', 'T'), (158, '-', 'G'), (159, '-', 'G'), (160, '-', 'T'), (161, '-', 'C'), (162, '-', 'A'), (163, '-', 'A'), (164, '-', 'C'), (165, '-', 'C'), (166, '-', 'A'), (167, '-', 'C'), (168, '-', 'C'), (169, '-', 'G'), (170, '-', 'C'), (171, '-', 'G'), (172, '-', 'G'), (173, '-', 'T'), (174, '-', 'C'), (175, '-', 'T'), (176, '-', 'C'), (177, '-', 'A'), (178, '-', 'G'), (179, '-', 'T'), (180, '-', 'G'), (181, '-', 'G'), (182, '-', 'T'), (183, '-', 'G'), (184, '-', 'T'), (185, '-', 'A'), (186, '-', 'C'), (187, '-', 'G'), (188, '-', 'G'), (189, '-', 'T'), (190, '-', 'A'), (191, '-', 'C'), (192, '-', 'A'), (193, '-', 'A'), (194, '-', 'A'), (195, '-', 'C'), (196, '-', 'C'), (197, '-', 'T')]\n",
      "Read 1779 alleles with 14469 reads\n",
      "Got 655 alleles with deletions\n",
      "Got 6106 perfect T alleles\n"
     ]
    }
   ],
   "source": [
    "print('Processing ' + crispresso_output_folder)\n",
    "crispresso2_info = CRISPRessoShared.load_crispresso_info(crispresso_output_folder)\n",
    "\n",
    "num_changes_to_check = 2\n",
    "\n",
    "if 'Comp' not in crispresso2_info['results']['refs']:\n",
    "    raise Exception('Reference sequence for reference \"Comp\" not found - was a reference sequence named \"Comp\" provided?')\n",
    "\n",
    "if not twin_seq:\n",
    "    raise Exception('Reference sequence for reference \"TwinPE\" not found - was a reference sequence named \"TwinPE\" provided?')\n",
    "\n",
    "comp_ref_seq  = crispresso2_info['results']['refs']['Comp']['sequence']\n",
    "comp_ref_name = crispresso2_info['results']['ref_names'][0]\n",
    "\n",
    "# Create alignments using the same scoring parameters CRISPResso2\n",
    "# used for read alignments to ensure consistent gap behavior.\n",
    "aln_gap_incentive = crispresso2_info['results']['refs'][comp_ref_name]['gap_incentive']\n",
    "aln_gap_open_arg = crispresso2_info['running_info']['args'].needleman_wunsch_gap_open\n",
    "aln_gap_extend_arg = crispresso2_info['running_info']['args'].needleman_wunsch_gap_extend\n",
    "\n",
    "aln_matrix_loc = crispresso2_info['running_info']['args'].needleman_wunsch_aln_matrix_loc\n",
    "if aln_matrix_loc == 'EDNAFULL':\n",
    "    aln_matrix = CRISPResso2Align.make_matrix()\n",
    "else:\n",
    "    if not os.path.exists(aln_matrix_loc):\n",
    "        raise Exception('Alignment matrix file not found at ' + aln_matrix_loc)\n",
    "    aln_matrix = CRISPResso2Align.read_matrix(aln_matrix_loc)\n",
    "\n",
    "# Align TwinPE to Compound Ref Sequence\n",
    "twin_aln_seq, comp_aln_seq_pe, aln_score = CRISPResso2Align.global_align(\n",
    "    twin_seq, \n",
    "    comp_ref_seq, \n",
    "    matrix=aln_matrix,\n",
    "    gap_incentive=aln_gap_incentive,\n",
    "    gap_open=aln_gap_open_arg,\n",
    "    gap_extend=aln_gap_extend_arg\n",
    ")\n",
    "\n",
    "# Align WT to Compound Ref Sequence\n",
    "wt_aln_seq, comp_aln_seq_wt, aln_score = CRISPResso2Align.global_align(\n",
    "    wt_seq, \n",
    "    comp_ref_seq, \n",
    "    matrix=aln_matrix,\n",
    "    gap_incentive=aln_gap_incentive,\n",
    "    gap_open=aln_gap_open_arg,\n",
    "    gap_extend=aln_gap_extend_arg\n",
    ")\n",
    "\n",
    "# Alignment debugging check - may need to change bp_changes_arr range if alignments differ (use longest?)\n",
    "if len(comp_aln_seq_pe) != len(comp_aln_seq_wt):\n",
    "    raise Exception('Compound ref alignments to TwinPE and WT have different lengths!')\n",
    "\n",
    "# Determine deletion and insertion regions in WT and TwinPE refs - these are the base changes\n",
    "bp_changes_arr = []\n",
    "del_start = del_end = ins_start = ins_end = None\n",
    "for idx in range(len(comp_aln_seq_pe)):\n",
    "    wt_base = wt_aln_seq[idx]\n",
    "    twin_base = twin_aln_seq[idx]\n",
    "    if wt_base != '-' and twin_base == '-':\n",
    "        bp_changes_arr.append((idx, wt_base, twin_base))\n",
    "        if del_start is None:\n",
    "            del_start = idx\n",
    "        del_end = idx\n",
    "    elif wt_base == '-' and twin_base != '-':\n",
    "        bp_changes_arr.append((idx, wt_base, twin_base))\n",
    "        if ins_start is None:\n",
    "            ins_start = idx\n",
    "        ins_end = idx\n",
    "del_region_coords = (del_start, del_end)\n",
    "ins_region_coords = (ins_start, ins_end)\n",
    "\n",
    "# Debugging checks\n",
    "del_region_len = del_end - del_start + 1\n",
    "ins_region_len = ins_end - ins_start + 1\n",
    "if len(bp_changes_arr) != del_region_len + ins_region_len:\n",
    "    raise Exception('Length of bp_changes_arr does not equal sum of del and ins region lengths.')\n",
    "\n",
    "print(f'Found {str(len(bp_changes_arr))} base changes (via compound alignment): {str(bp_changes_arr)}')\n",
    "\n",
    "# Map Compound ref bases to WT and TwinPE ref bases\n",
    "wt_map = get_refpos_values(comp_aln_seq_wt, wt_aln_seq)\n",
    "pe_map = get_refpos_values(comp_aln_seq_pe, twin_aln_seq)\n",
    "\n",
    "# Load allele frequency table\n",
    "z = zipfile.ZipFile(os.path.join(crispresso_output_folder, crispresso2_info['running_info']['allele_frequency_table_zip_filename']))\n",
    "zf = z.open(crispresso2_info['running_info']['allele_frequency_table_filename'])\n",
    "df_alleles = pd.read_csv(zf,sep=\"\\t\")\n",
    "\n",
    "nuc_bases = ['A','C','G','T']\n",
    "\n",
    "from_left_all_edit_counts = [0] * ins_region_len\n",
    "from_right_all_edit_counts = [0] * ins_region_len\n",
    "\n",
    "from_left_all_edit_counts_no_deletions = [0] * ins_region_len\n",
    "from_left_all_edit_counts_with_deletions = [0] * ins_region_len\n",
    "from_right_all_edit_counts_no_deletions = [0] * ins_region_len\n",
    "from_right_all_edit_counts_with_deletions = [0] * ins_region_len\n",
    "edit_counts_no_deletions = [0] * ins_region_len\n",
    "edit_counts_with_deletions = [0] * ins_region_len\n",
    "\n",
    "allele_counts = {} # e.g. TTRTT > 100\n",
    "allele_categories = {} # e.g. TTRTT > Left_flap\n",
    "\n",
    "edit_counts = [0] * ins_region_len\n",
    "deletion_counts = [0] * ins_region_len\n",
    "insertion_counts = [0] * ins_region_len\n",
    "perfect_T_count = 0\n",
    "\n",
    "total_alleles = 0\n",
    "total_alleles_reads = 0\n",
    "total_alleles_deletions_reads = 0\n",
    "\n",
    "# cat_wt_count = 0\n",
    "# cat_wt_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_wt_indel_count = 0\n",
    "# cat_wt_indel_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_left_flap_count = 0\n",
    "# cat_left_flap_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_right_flap_count = 0\n",
    "# cat_right_flap_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_perfect_pe_count = 0\n",
    "# cat_perfect_pe_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_imperfect_pe_count = 0\n",
    "# cat_imperfect_pe_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "# cat_pe_indels_count = 0\n",
    "# cat_pe_indels_count_arr = [0] * len(bp_changes_arr)\n",
    "\n",
    "cat_wt_count = 0\n",
    "cat_wt_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_wt_indel_count = 0\n",
    "cat_wt_indel_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_left_flap_count = 0\n",
    "cat_left_flap_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_right_flap_count = 0\n",
    "cat_right_flap_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_perfect_pe_count = 0\n",
    "cat_perfect_pe_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_imperfect_pe_count = 0\n",
    "cat_imperfect_pe_count_arr = [0] * ins_region_len\n",
    "\n",
    "cat_pe_indels_count = 0\n",
    "cat_pe_indels_count_arr = [0] * ins_region_len\n",
    "\n",
    "has_indel_in_match_arr_count = 0\n",
    "has_any_indel_byproduct_count = 0\n",
    "\n",
    "# iterate all alleles in input allele table\n",
    "for idx, allele in df_alleles.iterrows():\n",
    "    total_alleles += 1\n",
    "    total_alleles_reads += allele['#Reads']\n",
    "\n",
    "    comp_aln_seq_read = allele.Reference_Sequence\n",
    "    read_aln_seq = allele.Aligned_Sequence\n",
    "\n",
    "    # Map this reads aligned bases to Compound ref aligned bases \n",
    "    read_map = get_refpos_values(comp_aln_seq_read, read_aln_seq)\n",
    "\n",
    "    # Map this alleles match_arr index to ref_seq index\n",
    "    match_arr_to_comp_base_index = {}\n",
    "    # Build an array showing how this allele matches Comp vs TwinPE at each changed position.\n",
    "    # 'W' = matches WT base\n",
    "    # 'T' = matches TwinPE base\n",
    "    # 'D' = non-programmed deletion\n",
    "    # 'I' = non-programmed insertion\n",
    "    # 'S' = non-programmed substitution\n",
    "    # 'N' = ambiguous base\n",
    "    # Build an array for both the deletion and inserted regions (bp_changes_arr)\n",
    "    match_arr = []\n",
    "    for ind, (comp_ind, wt_base, twin_base) in enumerate(bp_changes_arr):\n",
    "        base_at_pos = read_map.get(comp_ind, \"\")\n",
    "        if base_at_pos == wt_base:\n",
    "            match_arr.append(\"W\")\n",
    "        elif base_at_pos == twin_base:\n",
    "            match_arr.append(\"T\")\n",
    "        elif base_at_pos == \"-\":\n",
    "            match_arr.append(\"D\")\n",
    "        # include insertions in PE_indel category if it still has all the edits?\n",
    "        # elif len(base_at_pos) > 1:\n",
    "        elif len(base_at_pos) > 1 and base_at_pos[0] == twin_base:\n",
    "            match_arr.append(\"T\")\n",
    "            # match_arr.append(\"I\")\n",
    "        elif base_at_pos in nuc_bases and base_at_pos != wt_base and base_at_pos != twin_base:\n",
    "            match_arr.append(\"S\")\n",
    "        else:\n",
    "            match_arr.append(\"N\")\n",
    "        match_arr_to_comp_base_index[ind] = comp_ind\n",
    "\n",
    "    # Array for only the sequence that is deleted by TwinPE\n",
    "    del_match_arr = match_arr[:del_end-del_start+1]\n",
    "    # Array for only the sequence that is inserted by TwinPE\n",
    "    ins_match_arr = match_arr[del_end-del_start+1:]\n",
    "\n",
    "    # # Build an array for both the deletion and inserted regions\n",
    "    # match_arr = [0] * len(bp_changes_arr)\n",
    "    # for ind, (comp_ind, wt_base, twin_base) in enumerate(bp_changes_arr):\n",
    "    #     base_at_pos = read_map.get(comp_ind, \"\")\n",
    "    #     if base_at_pos == wt_base:\n",
    "    #         match_arr[ind] = \"W\"\n",
    "    #     elif base_at_pos == twin_base:\n",
    "    #         match_arr[ind] = \"T\"\n",
    "    #     elif base_at_pos == \"-\":\n",
    "    #         match_arr[ind] = \"D\"\n",
    "    #     elif len(base_at_pos) > 1:\n",
    "    #         match_arr[ind] = \"I\"\n",
    "    #     else:\n",
    "    #         match_arr[ind] = \"N\"\n",
    "    #     match_arr_to_ref_base_index[ind] = comp_ind\n",
    "    \n",
    "    # # indels anywhere in the allele\n",
    "    # has_any_indel = False\n",
    "    # if allele.all_deletion_positions != '[]': # need to run with detailed allele output\n",
    "    #     has_any_indel = True\n",
    "    # if allele.all_insertion_positions != '[]':\n",
    "    #     has_any_indel = True\n",
    "\n",
    "    # # indels within quantification window\n",
    "    # has_indel_guide = False\n",
    "    # if allele.n_deleted > 0:\n",
    "    #     has_indel_guide = True\n",
    "    # if allele.n_inserted > 0:\n",
    "    #     has_indel_guide = True\n",
    "\n",
    "    # has_indel = has_indel_guide\n",
    "\n",
    "    # check match_arr for indels\n",
    "    has_indel_in_match_arr = any(m in [\"I\", \"D\"] for m in match_arr)\n",
    "    if has_indel_in_match_arr:\n",
    "        has_indel_in_match_arr_count += 1\n",
    "    # check allele table for indels anywhere\n",
    "    # check crispresso insertion list first\n",
    "    has_any_indel_byproduct = False\n",
    "    if allele.all_insertion_left_positions != \"[]\":\n",
    "        has_any_indel_byproduct = True\n",
    "    # check crispresso deletion list if needed - if deletion idx is not within programmed indel region\n",
    "    else:\n",
    "        edit_range = range(del_start, ins_end + 1)\n",
    "        all_del_pos = [int(x) for x in allele.all_deletion_positions.strip(\"[]\").split(\",\")]\n",
    "        for del_ind in all_del_pos:\n",
    "            if del_ind not in edit_range:\n",
    "                has_any_indel_byproduct = True\n",
    "                break\n",
    "    if has_any_indel_byproduct:\n",
    "        has_any_indel_byproduct_count += 1\n",
    "    \n",
    "    # has_indel = False\n",
    "    # if has_any_indel_byproduct or has_indel_in_match_arr:\n",
    "    has_indel = has_any_indel_byproduct\n",
    "    if has_indel:\n",
    "        total_alleles_deletions_reads += allele['#Reads']\n",
    "\n",
    "    match_arr_str = \"\\t\".join(match_arr) + \"\\t\" + str(has_indel)\n",
    "    if match_arr_str not in allele_counts:\n",
    "        allele_counts[match_arr_str] = 0\n",
    "    allele_counts[match_arr_str] += allele['#Reads']\n",
    "\n",
    "    # Define PE sites -> every position that differs between WT and PE\n",
    "\n",
    "    # 1st break:\n",
    "    # WT vs Prime edited\n",
    "\n",
    "    # If every PE site is PE \n",
    "    # If it has indels -> PE indel (indel anywhere - including outside of guide windows)\n",
    "    # if not -> perfect PE\n",
    "\n",
    "    # If every PE site is WT\n",
    "    # If it has indels - >WT indel (anywhere, even outside)\n",
    "    # if not, -> WT\n",
    "\n",
    "    # If left two PE sites are PE:\n",
    "    # Left flap\n",
    "    # If right two PE sites are PE:\n",
    "    # Right flap\n",
    "    # Otherwise, \n",
    "    # Imperfect PE\n",
    "\n",
    "    total_TE_count = match_arr.count(\"T\")\n",
    "    has_all_TE = (total_TE_count == len(match_arr))\n",
    "\n",
    "    total_WT_count = match_arr.count(\"W\")\n",
    "    has_all_WT = (total_WT_count == len(match_arr))\n",
    "\n",
    "    # ins_TE_count = ins_match_arr.count(\"T\")\n",
    "    # is_imperfect_PE = (ins_TE_count >= num_changes_to_check and ins_TE_count < len(ins_match_arr))\n",
    "\n",
    "    # define left/right flap booleans\n",
    "    is_left_flap = False\n",
    "    is_right_flap = False\n",
    "    if len(ins_match_arr) >= num_changes_to_check:\n",
    "        is_left_flap = all(m == \"T\" for m in ins_match_arr[:num_changes_to_check])\n",
    "        is_right_flap = all(m == \"T\" for m in ins_match_arr[-num_changes_to_check:])\n",
    "        # is_left_flap = all(m == \"T\" for m in match_arr[del_end-del_start+1:del_end-del_start+1+num_changes_to_check])\n",
    "        # is_right_flap = all(m == \"T\" for m in match_arr[-num_changes_to_check:])\n",
    "\n",
    "    # is_left_flap = True\n",
    "    # for i in range(num_changes_to_check):\n",
    "    #     if len(match_arr) > 1:\n",
    "    #         if match_arr[i] != 'T':\n",
    "    #             is_left_flap = False\n",
    "\n",
    "    # is_right_flap = True\n",
    "    # for i in range(num_changes_to_check):\n",
    "    #     if len(match_arr) > 1:\n",
    "    #         if match_arr[-1-i] != 'T':\n",
    "    #             is_right_flap = False\n",
    "\n",
    "    # TE_count = 0\n",
    "    # for match_val in match_arr:\n",
    "    #     if match_val == 'T':\n",
    "    #         TE_count += 1\n",
    "\n",
    "    # is_perfect_PE = False\n",
    "    # if TE_count == len(match_arr):\n",
    "    #     is_perfect_PE = True\n",
    "    \n",
    "    # is_imperfect_PE = False\n",
    "    # if TE_count > num_changes_to_check:\n",
    "    #     is_imperfect_PE = True\n",
    "\n",
    "    category = \"NA\"\n",
    "    if has_all_TE and has_indel:\n",
    "        category = \"PE_Indel\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_pe_indels_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_pe_indels_count_arr[pos_idx] += allele['#Reads']\n",
    "    \n",
    "    elif has_all_TE:\n",
    "        category = \"Perfect_PE\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_perfect_pe_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_perfect_pe_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    elif has_all_WT and has_indel:\n",
    "        category = \"WT_Indel\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_wt_indel_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_wt_indel_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    elif has_all_WT:\n",
    "        category = \"WT\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_wt_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_wt_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    elif is_left_flap and not is_right_flap:\n",
    "        category = \"Left_Flap\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_left_flap_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_left_flap_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    elif is_right_flap and not is_left_flap:\n",
    "        category = \"Right_Flap\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_right_flap_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_right_flap_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    else:\n",
    "        category = \"Imperfect_PE\"\n",
    "        df_alleles.at[idx,'Category'] = category\n",
    "        cat_imperfect_pe_count += allele['#Reads']\n",
    "        for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "            if match == 'T':\n",
    "                cat_imperfect_pe_count_arr[pos_idx] += allele['#Reads']\n",
    "\n",
    "    if match_arr_str in allele_categories and allele_categories[match_arr_str] != category:\n",
    "        raise Exception('conflicting categories for ' + match_arr_str + ' ' + category + ' vs ' + allele_categories[match_arr_str])\n",
    "    allele_categories[match_arr_str] = category\n",
    "\n",
    "    for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "        if match == 'T':\n",
    "            from_left_all_edit_counts[pos_idx] += allele['#Reads']\n",
    "\n",
    "            if has_indel:\n",
    "                from_left_all_edit_counts_with_deletions[pos_idx] += allele['#Reads']\n",
    "            else:\n",
    "                from_left_all_edit_counts_no_deletions[pos_idx] += allele['#Reads']\n",
    "\n",
    "        else:\n",
    "            break\n",
    "\n",
    "    for pos_idx, match in zip(reversed(range(len(ins_match_arr))), reversed(ins_match_arr)):\n",
    "        if match == 'T':\n",
    "            from_right_all_edit_counts[pos_idx] += allele['#Reads']\n",
    "\n",
    "            if has_indel:\n",
    "                from_right_all_edit_counts_with_deletions[pos_idx] += allele['#Reads']\n",
    "            else:\n",
    "                from_right_all_edit_counts_no_deletions[pos_idx] += allele['#Reads']\n",
    "        else:\n",
    "            break\n",
    "\n",
    "    for pos_idx, match in zip(range(len(ins_match_arr)), ins_match_arr):\n",
    "        if match == 'T':\n",
    "            edit_counts[pos_idx] += allele['#Reads']\n",
    "            if has_indel:\n",
    "                edit_counts_with_deletions[pos_idx] += allele['#Reads']\n",
    "            else:\n",
    "                edit_counts_no_deletions[pos_idx] += allele['#Reads']\n",
    "        if match == 'D':\n",
    "            deletion_counts[pos_idx] += allele['#Reads']\n",
    "        if match == \"I\":\n",
    "            insertion_counts[pos_idx] += allele[\"#Reads\"]\n",
    "    if ins_match_arr == ['T']*len(ins_match_arr):\n",
    "        perfect_T_count += allele['#Reads']\n",
    "\n",
    "    #### DEBUG ####\n",
    "    # # if read_aln == 'CGCCGGGAACTGCCGCTGGCCCCCCACCGCC---AGGTTTGTCTGGTCAACCACCGCGGT-------------------CTCAGTGGTGTACGGTACAAACCTC------CACTTCCA---GGAGTCGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTGGGAAGTCCGCAGCTCCGAGGCGCCCAGTGGAAATCGCCAGATGAGGGCCTCCTC':\n",
    "    # if 'CGCCGGGAACTGCCGCTGGCCCCCCACCGCC---------------------------------------------------------------------------------------------------------------------AGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTGGGAAGTCCGCAGCT----GGCCCCCCACCGCCAGGTTTCGCCAGATGAGGGCCTCCTC' in read_aln_seq:\n",
    "    #     print(f\"{allele['#Reads'] =  }\")\n",
    "    #     # print(f\"{allele['Category'] =  }\")\n",
    "    #     print(f\"{category =          }\")\n",
    "    #     print(f\"{wt_aln_seq =        }\")\n",
    "    #     print(f\"{twin_aln_seq =      }\")\n",
    "    #     print(f\"{comp_aln_seq_pe =   }\")\n",
    "    #     print(f\"{comp_aln_seq_read = }\")\n",
    "    #     print(f\"{read_aln_seq =      }\")\n",
    "    #     print(f\"{all_del_pos =                            }\")\n",
    "    #     print(f\"{allele['all_deletion_positions'] =       }\")\n",
    "    #     print(f\"{allele['all_insertion_left_positions'] = }\")\n",
    "    #     print(f\"{bp_changes_arr = }\")\n",
    "    #     print(f\"{match_arr =      }\")\n",
    "    #     print(f\"{del_match_arr =  }\")\n",
    "    #     print(f\"{ins_match_arr =  }\")\n",
    "    #     print(f\"{del_region_coords = }\")\n",
    "    #     print(f\"{ins_region_coords = }\")\n",
    "    #     print(f\"{match_arr_to_comp_base_index = }\")\n",
    "    #     print(f\"{read_map = }\")\n",
    "    #     print(f\"{wt_map =   }\")\n",
    "    #     print(f\"{pe_map =   }\")\n",
    "    #     print(f\"{has_indel =  }\")\n",
    "    #     print(f\"{has_all_WT = }\")\n",
    "    #     print(f\"{has_all_TE = }\")\n",
    "    #     # print(allele.all_insertion_positions)\n",
    "    #     # print(allele.all_insertion_left_positions)\n",
    "\n",
    "imperfect_from_left_all_edit_counts = [x-perfect_T_count for x in from_left_all_edit_counts] #these are reads that start with a twin-edited, but are not completely twin-edited. Note that from_left_all_edit_counts includes perfect matches\n",
    "imperfect_from_right_all_edit_counts = [x-perfect_T_count for x in from_right_all_edit_counts]\n",
    "perfect_edit_counts = [perfect_T_count] * ins_region_len\n",
    "total_counts = [total_alleles_reads] * ins_region_len\n",
    "\n",
    "# with open('01_classify_reads_7cat.ipynb.allele_counts.txt','w') as fout:\n",
    "#     sorted_allele_counts = sorted(allele_counts.keys(), key=lambda x: allele_counts[x], reverse=True)\n",
    "#     fout.write(\"\\t\".join([str(x) for x in bp_changes_arr]) + '\\thas_indel\\tcount\\n')\n",
    "#     for allele_str in sorted_allele_counts:\n",
    "#         fout.write(allele_str + '\\t' + str(allele_counts[allele_str]) + '\\n')\n",
    "        \n",
    "# with open('01_classify_reads_7cat.ipynb.arrays.txt','w') as fout:\n",
    "#     fout.write('Class\\t' + \"\\t\".join([str(x) for x in bp_changes_arr]) + '\\n')\n",
    "#     fout.write('total_counts\\t' + \"\\t\".join([str(x) for x in total_counts]) + '\\n')\n",
    "#     fout.write('edit_counts\\t' + \"\\t\".join([str(x) for x in edit_counts]) + '\\n')\n",
    "#     fout.write('perfect_edit_counts\\t' + \"\\t\".join([str(x) for x in perfect_edit_counts]) + '\\n')\n",
    "#     fout.write('from_left_all_edit_counts\\t' + \"\\t\".join([str(x) for x in from_left_all_edit_counts]) + '\\n')\n",
    "#     fout.write('from_right_all_edit_counts\\t' + \"\\t\".join([str(x) for x in from_right_all_edit_counts]) + '\\n')\n",
    "#     fout.write('imperfect_from_left_all_edit_counts\\t' + \"\\t\".join([str(x) for x in imperfect_from_left_all_edit_counts]) + '\\n')\n",
    "#     fout.write('imperfect_from_right_all_edit_counts\\t' + \"\\t\".join([str(x) for x in imperfect_from_right_all_edit_counts]) + '\\n')\n",
    "#     fout.write('deletion_counts\\t' + \"\\t\".join([str(x) for x in deletion_counts]) + '\\n')\n",
    "\n",
    "# with open('01_classify_reads_7cat.ipynb.counts.txt','w') as fout:\n",
    "#     fout.write(\"\\t\".join(['WT','WT_indels','left_flap','right_flap','perfect_PE','imperfect_PE','PE_indels']) + '\\n')\n",
    "#     fout.write(\"\\t\".join([str(x) for x in [cat_wt_count, cat_wt_indel_count, cat_left_flap_count, cat_right_flap_count, cat_perfect_pe_count, cat_imperfect_pe_count, cat_pe_indels_count]]) + '\\n')\n",
    "\n",
    "\n",
    "print('Read ' + str(total_alleles) + ' alleles with ' + str(total_alleles_reads) + ' reads')\n",
    "print('Got ' + str(total_alleles_deletions_reads) + ' alleles with deletions')\n",
    "print('Got ' + str(perfect_T_count) + ' perfect T alleles')\n",
    "# print(f\"{has_indel_in_match_arr_count = }\")\n",
    "# print(f\"{has_any_indel_byproduct_count = }\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b841eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tot_reads = 0\n",
    "dff = df_alleles[df_alleles['Category'] == 'PE_indels']\n",
    "for i, row in dff.iterrows():\n",
    "    if 'CGCCGGGAACTGCCGCTGGCCCCCCACCGCC---------------------------------------------------------------------------------------------------------------------AGGTTTGTCTGGTCAACCACCGCGGTCTCAGTGGTGTACGGTACAAACCTCCACTTCCAGGAGTCGCTGTGCCCCGATGCACACTG' in row['Aligned_Sequence']:\n",
    "        print(i, row['#Reads'], row['Aligned_Sequence'])        \n",
    "        tot_reads += row['#Reads']\n",
    "print(f'Total PE_indels reads with specific sequence: {tot_reads}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbf00759",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in dff.sort_values(by='#Reads', ascending=False).head(25).iterrows():\n",
    "    print(i, row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2918927c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(wt_aln_seq)\n",
    "# print(wt_ref_aln_seq)\n",
    "# print(ref_aln_seq)\n",
    "# print(twin_aln_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c0a8896",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df_alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "022b5060",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1779"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_alleles.Category.value_counts().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f496acde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Category\n",
       "Imperfect_PE    640\n",
       "WT              402\n",
       "Right_Flap      305\n",
       "Perfect_PE      283\n",
       "PE_Indel         80\n",
       "WT_Indel         53\n",
       "Left_Flap        16\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_alleles.Category.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65805d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_alleles.Category.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acd30a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sss = 'CCCAGCCAAACTTGTCAACCAGTATCCCGGCAGACTGAGCACGGGAGTGCAGGTGGAAACCATCTCCCCAGGAGACGGGCGCACCTTCCCCAAGCGCGGCCAGACCTGCGTGGTCCAGTATCCCGGAAATTTCTTTCCATCTTCAAGCATCCCGGTGTAGTGCACCACGCAGGTCTGGCCGCGCTTGGGGAAGGTTGGGCCCCAAGGATTGACCCAGGCCAGGGCTGGAGAAGCAGAAAAAAAGCATCAAGCCTACAAATGCATGCTTACTTGCAGCAGAAATAGACTAATTGCATGGGCGTTTCCCTGGGATCCCTGTCTCCAGGCTTCTCTCATCCATGCCT'\n",
    "# ins_108_A ='ACCATCTCCCCAGGAGACGGGCGCACCTTCCCCAAGCGCGGCCAGACCTGCGTGGTCCAGTATCCCGGAAATTTCTTTCCATCTTCAAGCATCCCGGTGTAGTGCACC'\n",
    "# ins_108_B ='CCATCTCCCCAGGAGACGGGCGCACCTTCCCCAAGCGCGGCCAGACCTGCGTGGTCCAGTATCCCGGAAATTTCTTTCCATCTTCAAGCATCCCGGTGTAGTGCACCA'\n",
    "for i, row in df_alleles.sort_values(by='#Reads', ascending=False).head(25).iterrows():\n",
    "    print(row['#Reads'])\n",
    "    print(row['Reference_Sequence'])\n",
    "    # print(len(row['Reference_Sequence']))\n",
    "    print(row['Aligned_Sequence'])\n",
    "    print(row['Category'])\n",
    "    print(row.all_insertion_left_positions)\n",
    "    print(row.all_insertion_positions)\n",
    "    print(row.all_deletion_positions)\n",
    "    # print(len(row['Aligned_Sequence']))\n",
    "    # if pe3_108_ref_seq_substring in row['Aligned_Sequence'].replace('-',''):\n",
    "    # print(df_alleles.sort_values(by='#Reads', ascending=False).iloc[i]['Reference_Sequence'])\n",
    "    # print(df_alleles.sort_values(by='#Reads', ascending=False).iloc[i]['Aligned_Sequence'])\n",
    "    # AAATTTCTTTCCATCTTCAAGCATCCCGGTGTAGTGCACCACGCAGGTCTGGCCGCGCTTGGGGAAGGTGCGCCCGTCTCCTGGGGAGATGGTTTCCACCTGCACTCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ec623d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def setBarMatplotlibDefaults():\n",
    "    # matplotlib.use('AGG')\n",
    "    # font = {'size': 22}\n",
    "    # matplotlib.rc('font', **font)\n",
    "    matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "    matplotlib.rcParams['ps.fonttype'] = 42\n",
    "    matplotlib.rcParams[\"font.sans-serif\"] = [\"Arial\", \"Liberation Sans\", \"Bitstream Vera Sans\"]\n",
    "    matplotlib.rcParams[\"font.family\"] = \"sans-serif\"\n",
    "    matplotlib.rcParams['axes.facecolor'] = 'white'\n",
    "    # plt.ioff()\n",
    "\n",
    "setBarMatplotlibDefaults() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8fff97f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_root = 'figures/02_final_classify_reads_7cat/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261aaad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indices for the x-axis\n",
    "indices = np.arange(len(edit_counts))\n",
    "\n",
    "# Plotting\n",
    "fig, ax = plt.subplots(figsize=(14, 6)) \n",
    "\n",
    "# Stacked bar plot\n",
    "ax.bar(indices, cat_perfect_pe_count_arr, label='Perfect PE')\n",
    "bottom_so_far = cat_perfect_pe_count_arr\n",
    "\n",
    "ax.bar(indices, cat_left_flap_count_arr, label='Left Flap', bottom=bottom_so_far)\n",
    "bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_left_flap_count_arr)]\n",
    "\n",
    "ax.bar(indices, cat_right_flap_count_arr, label='Right Flap', bottom=bottom_so_far)\n",
    "bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_right_flap_count_arr)]\n",
    "\n",
    "ax.bar(indices, cat_imperfect_pe_count_arr, label='Imperfect PE', bottom=bottom_so_far)\n",
    "bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_imperfect_pe_count_arr)]\n",
    "\n",
    "ax.bar(indices, cat_pe_indels_count_arr, label='PE Indels', bottom=bottom_so_far)\n",
    "bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_pe_indels_count_arr)]\n",
    "\n",
    "# By definition these cannot have any 'successful twin edits' so not including\n",
    "# ax.bar(indices, cat_wt_indel_count_arr, label='WT Indels', bottom=bottom_so_far)\n",
    "# bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_wt_indel_count_arr)]\n",
    "\n",
    "# ax.bar(indices, cat_wt_count_arr, label='WT', bottom=bottom_so_far)\n",
    "# bottom_so_far = [x + y for x, y in zip(bottom_so_far, cat_wt_count_arr)]\n",
    "\n",
    "# Adding labels and title\n",
    "ax.set_xlabel('Compound Reference Index')\n",
    "ax.set_ylabel('Read Counts')\n",
    "ax.set_title('Stacked Successful Twin Edits by Category')\n",
    "\n",
    "ax.set_xticks(indices)\n",
    "ax.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "# Dynamic? >35 x-tick = font 8 & rotate 45\n",
    "ax.tick_params(axis='x', labelsize=8)\n",
    "ax.legend()\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.savefig(fig_root + '.successful_twin_edit_counts_by_category.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n",
    "\n",
    "# # Display or save\n",
    "# if fig_filename_root is None:\n",
    "#     plt.show()\n",
    "# else:\n",
    "#     fig.savefig(fig_filename_root+'.pdf', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "#     if SAVE_ALSO_PNG:\n",
    "#         fig.savefig(fig_filename_root+'.png', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "# plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af16c1fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "deletion_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d389560",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indices for the x-axis\n",
    "indices = np.arange(len(edit_counts))\n",
    "\n",
    "# Plotting\n",
    "fig, ax = plt.subplots(figsize=(14, 6)) \n",
    "\n",
    "# Stacked bar plot\n",
    "ax.bar(indices, total_counts, label='Total Counts')\n",
    "ax.bar(indices, edit_counts, label='Total TE Edit Counts')\n",
    "ax.bar(indices, from_right_all_edit_counts, label='From Right All TE Edit Counts')\n",
    "ax.bar(indices, from_left_all_edit_counts, label='From Left All TE Edit Counts')\n",
    "ax.bar(indices, perfect_edit_counts, label='Perfect TE Edit Counts')\n",
    "ax.bar(indices, deletion_counts, label='Deletion Counts')\n",
    "ax.bar(indices, insertion_counts, label='Insertion Counts')\n",
    "\n",
    "# Adding labels and title\n",
    "ax.set_xlabel('Compound Reference Index')\n",
    "ax.set_ylabel('Read Counts')\n",
    "ax.set_title('Successful Twin Edit Counts - All Alleles')\n",
    "\n",
    "ax.set_xticks(indices)\n",
    "ax.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "# Dynamic? >35 x-tick = font 8 & rotate 45\n",
    "ax.tick_params(axis='x', labelsize=8)\n",
    "ax.legend()\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.savefig(fig_root + '.successful_twin_edit_counts_total.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91d2d837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indices for the x-axis\n",
    "indices = np.arange(len(edit_counts))\n",
    "\n",
    "# Plotting\n",
    "fig, ax = plt.subplots(figsize=(14, 6)) \n",
    "\n",
    "# Stacked bar plot\n",
    "ax.bar(indices, edit_counts, label='Total TE Edit Counts')\n",
    "ax.bar(indices, from_right_all_edit_counts, label='From Right All TE Edit Counts')\n",
    "ax.bar(indices, from_left_all_edit_counts, label='From Left All TE Edit Counts')\n",
    "ax.bar(indices, perfect_edit_counts, label='Perfect TE Edit Counts')\n",
    "ax.bar(indices, deletion_counts, label='Deletion Counts')\n",
    "ax.bar(indices, insertion_counts, label='Insertion Counts')\n",
    "\n",
    "# Adding labels and title\n",
    "ax.set_xlabel('Compound Reference Index')\n",
    "ax.set_ylabel('Read Counts')\n",
    "ax.set_title('Successful Twin Edit Counts - Edited Alleles')\n",
    "\n",
    "ax.set_xticks(indices)\n",
    "ax.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "# Dynamic? >35 x-tick = font 8 & rotate 45\n",
    "ax.tick_params(axis='x', labelsize=8)\n",
    "ax.legend()\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.savefig(fig_root + '.successful_twin_edit_counts.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce79929",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indices for the x-axis\n",
    "indices = np.arange(len(edit_counts))\n",
    "\n",
    "# Plotting\n",
    "fig, (ax1, ax2, ax3) = plt.subplots(3,1,figsize=(14, 18)) \n",
    "\n",
    "# Stacked bar plot\n",
    "ax1.bar(indices, edit_counts, label='Total TE Edit Counts')\n",
    "ax1.bar(indices, edit_counts_with_deletions, label='Total TE Edit Counts (With indels)')\n",
    "\n",
    "# Adding labels and title\n",
    "ax1.set_xlabel('Compound Reference Index')\n",
    "ax1.set_ylabel('Read Counts')\n",
    "ax1.set_title('Successful Twin Edit Counts')\n",
    "\n",
    "ax1.set_xticks(indices)\n",
    "ax1.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "ax1.tick_params(axis='x', labelsize=8)\n",
    "ax1.legend()\n",
    "\n",
    "# Stacked bar plot\n",
    "ax2.bar(indices, from_left_all_edit_counts, label='From Left All TE Edit Counts')\n",
    "ax2.bar(indices, from_left_all_edit_counts_with_deletions, label='From Left All TE Edit Counts (With indels)')\n",
    "\n",
    "# Adding labels and title\n",
    "ax2.set_xlabel('Compound Reference Index')\n",
    "ax2.set_ylabel('Read Counts')\n",
    "ax2.set_title('Twin Edit From Left Counts')\n",
    "\n",
    "ax2.set_xticks(indices)\n",
    "ax2.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "ax2.tick_params(axis='x', labelsize=8)\n",
    "ax2.legend()\n",
    "\n",
    "\n",
    "# Stacked bar plot\n",
    "ax3.bar(indices, from_right_all_edit_counts, label='From Right All TE Edit Counts')\n",
    "ax3.bar(indices, from_right_all_edit_counts_with_deletions, label='From Right All TE Edit Counts (With indels)')\n",
    "\n",
    "# Adding labels and title\n",
    "ax3.set_xlabel('Compound Reference Index')\n",
    "ax3.set_ylabel('Read Counts')\n",
    "ax3.set_title('Twin Edit From Right Counts')\n",
    "\n",
    "ax3.set_xticks(indices)\n",
    "ax3.set_xticklabels([str(x) for x in range(ins_start, ins_end+1)], rotation=45, ha='center')\n",
    "# Dynamic? >35 x-tick = font 8 & rotate 45\n",
    "ax3.tick_params(axis='x', labelsize=8)\n",
    "ax3.legend()\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "# Display the plot\n",
    "plt.savefig(fig_root + '.successful_twin_edit_counts_triple.pdf')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d92ddb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def add_sgRNA_to_ax(ax,sgRNA_intervals,sgRNA_y_start,sgRNA_y_height,amp_len,x_offset=0,sgRNA_mismatches=None,sgRNA_names=None,sgRNA_rows=None,font_size=None,clip_on=True,label_at_zero=False):\n",
    "#     \"\"\"\n",
    "#     Draw one or more sgRNA annotations on a Matplotlib Axes that shares the x-axis\n",
    "#     (reference/amplicon coordinates) with a heatmap of alleles.\n",
    "\n",
    "#     Each sgRNA is represented by a semi-transparent rectangle spanning its start and\n",
    "#     end coordinates. Optional per-sgRNA mismatches are overlaid as red sub-rectangles,\n",
    "#     and optional names are placed to the immediate left of each sgRNA. When multiple\n",
    "#     sgRNAs overlap in x, they can be vertically stacked by assigning them to distinct\n",
    "#     \"rows\" to avoid visual overlap.\n",
    "\n",
    "#     Args:\n",
    "#         ax (matplotlib.axes.Axes): Target axes to draw sgRNA rectangles and labels.\n",
    "#         sgRNA_intervals (list[tuple[int,int]]): For each sgRNA, a (start, end) tuple\n",
    "#             in reference coordinates. Values can extend beyond [0, amp_len-1] and\n",
    "#             will be clipped to the plotable region. Start and end are inclusive.\n",
    "#         sgRNA_y_start (float): Baseline y coordinate for the first sgRNA row, in the\n",
    "#             data coordinate system of ax.\n",
    "#         sgRNA_y_height (float): Total vertical span available for all sgRNA rows. If\n",
    "#             multiple rows are used, each row's height is sgRNA_y_height / n_rows.\n",
    "#         amp_len (int): Length of the amplicon/reference plotted on the x-axis.\n",
    "#         x_offset (int, optional): Horizontal offset to apply to all sgRNA x positions\n",
    "#             when the plotted x coordinates are shifted relative to reference indexing.\n",
    "#             Defaults to 0.\n",
    "#         sgRNA_mismatches (list[list[int]] | None, optional): For each sgRNA, a list of\n",
    "#             0-based indices (relative to the sgRNA start) where mismatches should be\n",
    "#             highlighted in red. Defaults to None.\n",
    "#         sgRNA_names (list[str] | None, optional): Optional text labels for each sgRNA.\n",
    "#             If provided, the label is drawn to the left of the sgRNA span. Defaults to None.\n",
    "#         sgRNA_rows (list[int] | None, optional): Row index for each sgRNA used to\n",
    "#             vertically stack overlapping sgRNAs. 0 is the top-most row. If None, all\n",
    "#             sgRNAs are placed on row 0. Defaults to None.\n",
    "#         font_size (int | str | None, optional): Font size for sgRNA labels. If None,\n",
    "#             matplotlib.rcParams['font.size'] is used. Defaults to None.\n",
    "#         clip_on (bool, optional): Whether sgRNA patches/text should be clipped to\n",
    "#             the Axes' view limits. Set False to allow drawing outside the plot area\n",
    "#             (useful if labels should appear beyond the axes). Defaults to True.\n",
    "#         label_at_zero (bool, optional): If True and the left-most sgRNA would be\n",
    "#             labeled off-plot, anchor the label at x=0 instead. Defaults to False.\n",
    "\n",
    "#     Notes:\n",
    "#         - Intervals that extend beyond [0, amp_len-1] are visually capped with small\n",
    "#           white rectangles to indicate clipping.\n",
    "#         - The function only adds artists (patches/text) to the provided Axes and\n",
    "#           does not alter limits, ticks, or other axis properties.\n",
    "#     \"\"\"\n",
    "\n",
    "#     if font_size is None:\n",
    "#         font_size = matplotlib.rcParams['font.size']\n",
    "\n",
    "#     # If rows are not provided, draw all sgRNAs on a single row\n",
    "#     if sgRNA_rows is None:\n",
    "#         sgRNA_rows = [0]*len(sgRNA_intervals)\n",
    "#     max_sgRNA_row = max(sgRNA_rows)+1\n",
    "#     # Height allocated per row\n",
    "#     this_sgRNA_y_height = sgRNA_y_height/float(max_sgRNA_row)\n",
    "\n",
    "#     # Track left-most sgRNA to optionally place a generic label\n",
    "#     min_sgRNA_x = None\n",
    "#     label_left_sgRNA = True\n",
    "\n",
    "#     for idx, sgRNA_int in enumerate(sgRNA_intervals):\n",
    "#         # Clip interval to the drawable region\n",
    "#         this_sgRNA_start = max(0, sgRNA_int[0])\n",
    "#         this_sgRNA_end = min(sgRNA_int[1], amp_len - 1)\n",
    "#         if this_sgRNA_start > amp_len or this_sgRNA_end < 0:\n",
    "#             # Entire interval falls outside the region; skip\n",
    "#             continue\n",
    "\n",
    "#         # Compute the per-row baseline start\n",
    "#         this_sgRNA_y_start = sgRNA_y_start + this_sgRNA_y_height*sgRNA_rows[idx]\n",
    "\n",
    "#         # Main sgRNA span as a translucent rectangle\n",
    "#         ax.add_patch(\n",
    "#             patches.Rectangle(\n",
    "#                 (x_offset+this_sgRNA_start, this_sgRNA_y_start),\n",
    "#                 1+this_sgRNA_end-this_sgRNA_start,\n",
    "#                 this_sgRNA_y_height,\n",
    "#                 facecolor=(0, 0, 0, 0.15),\n",
    "#                 clip_on=clip_on\n",
    "#             )\n",
    "#         )\n",
    "\n",
    "#         # If clipped on the left, draw a small white cap to indicate truncation\n",
    "#         if (this_sgRNA_start) != sgRNA_int[0]:\n",
    "#             ax.add_patch(\n",
    "#                 patches.Rectangle(\n",
    "#                     (x_offset + 0.1+this_sgRNA_start, this_sgRNA_y_start),\n",
    "#                     0.1,\n",
    "#                     this_sgRNA_y_height,\n",
    "#                     facecolor='w',\n",
    "#                     clip_on=clip_on\n",
    "#                 )\n",
    "#             )\n",
    "#         # If clipped on the right, draw a white cap\n",
    "#         if this_sgRNA_end != sgRNA_int[1]:\n",
    "#             ax.add_patch(\n",
    "#                 patches.Rectangle(\n",
    "#                     (x_offset + 0.8+this_sgRNA_end, this_sgRNA_y_start),\n",
    "#                     0.1,\n",
    "#                     this_sgRNA_y_height,\n",
    "#                     facecolor='w',\n",
    "#                     clip_on=clip_on\n",
    "#                 )\n",
    "#             )\n",
    "\n",
    "#         # Optional per-sgRNA mismatches overlaid in red\n",
    "#         if sgRNA_mismatches is not None:\n",
    "#             this_sgRNA_mismatches = sgRNA_mismatches[idx]\n",
    "#             for mismatch in this_sgRNA_mismatches:\n",
    "#                 mismatch_plot_pos = sgRNA_int[0] + mismatch\n",
    "#                 # Only draw if mismatch falls within the drawable region\n",
    "#                 if mismatch_plot_pos > 0 and mismatch_plot_pos < amp_len - 1:\n",
    "#                     ax.add_patch(\n",
    "#                         patches.Rectangle(\n",
    "#                             (x_offset+ mismatch_plot_pos, this_sgRNA_y_start),\n",
    "#                             1,\n",
    "#                             this_sgRNA_y_height,\n",
    "#                             facecolor='r',\n",
    "#                             clip_on=clip_on\n",
    "#                         )\n",
    "#                     )\n",
    "\n",
    "#         # Track the minimum (left-most) start for optional generic labeling\n",
    "#         if not min_sgRNA_x:\n",
    "#             min_sgRNA_x = this_sgRNA_start\n",
    "#         if this_sgRNA_start < min_sgRNA_x:\n",
    "#             min_sgRNA_x = this_sgRNA_start\n",
    "\n",
    "#         # Optional per-sgRNA label\n",
    "#         if sgRNA_names is not None and idx < len(sgRNA_names) and sgRNA_names[idx] != \"\":\n",
    "#             if (label_at_zero and x_offset + this_sgRNA_start < len(sgRNA_names[idx])*0.66):\n",
    "#                 ax.text(\n",
    "#                     0,\n",
    "#                     this_sgRNA_y_start + this_sgRNA_y_height/2,\n",
    "#                     sgRNA_names[idx] + \" \",\n",
    "#                     horizontalalignment='left',\n",
    "#                     verticalalignment='center',\n",
    "#                     fontsize = font_size\n",
    "#                 )\n",
    "#             else:\n",
    "#                 ax.text(\n",
    "#                     x_offset+this_sgRNA_start,\n",
    "#                     this_sgRNA_y_start + this_sgRNA_y_height/2,\n",
    "#                     sgRNA_names[idx] + \" \",\n",
    "#                     horizontalalignment='right',\n",
    "#                     verticalalignment='center',\n",
    "#                     fontsize = font_size\n",
    "#                 )\n",
    "#             label_left_sgRNA = False  # At least one sgRNA has been explicitly labeled\n",
    "\n",
    "#     # If nothing was explicitly labeled, draw a generic label near the left-most sgRNA\n",
    "#     if min_sgRNA_x is not None and label_left_sgRNA:\n",
    "#         if (label_at_zero and x_offset + min_sgRNA_x < 5):\n",
    "#             ax.text(0, this_sgRNA_y_start + this_sgRNA_y_height/2, 'sgRNA ', horizontalalignment='left', verticalalignment='center', fontsize=font_size)\n",
    "#         else:\n",
    "#             ax.text(x_offset+min_sgRNA_x, this_sgRNA_y_start + this_sgRNA_y_height/2, 'sgRNA ', horizontalalignment='right', verticalalignment='center', fontsize=font_size)\n",
    "\n",
    "#### Modified function to allow right-side labels ####\n",
    "# // ...existing code...\n",
    "# def add_sgRNA_to_ax(ax,\n",
    "#                     sgRNA_intervals,\n",
    "#                     sgRNA_y_start,\n",
    "#                     sgRNA_y_height,\n",
    "#                     amp_len,\n",
    "#                     x_offset=0,\n",
    "#                     sgRNA_mismatches=None,\n",
    "#                     sgRNA_names=None,\n",
    "#                     sgRNA_rows=None,\n",
    "#                     font_size=None,\n",
    "#                     clip_on=True,\n",
    "#                     label_at_zero=False,\n",
    "#                     sgRNA_label_sides=None):\n",
    "#     \"\"\"\n",
    "#     Draw one or more sgRNA annotations on a Matplotlib Axes.\n",
    "\n",
    "#     Each sgRNA is a semi-transparent rectangle (start..end). Optional mismatches\n",
    "#     are red sub-rectangles. Labels can be placed per sgRNA either to the left\n",
    "#     or right of the rectangle via sgRNA_label_sides.\n",
    "\n",
    "#     Args:\n",
    "#         ax (matplotlib.axes.Axes)\n",
    "#         sgRNA_intervals (list[tuple[int,int]])\n",
    "#         sgRNA_y_start (float)\n",
    "#         sgRNA_y_height (float)\n",
    "#         amp_len (int)\n",
    "#         x_offset (int)\n",
    "#         sgRNA_mismatches (list[list[int]] | None)\n",
    "#         sgRNA_names (list[str] | None)\n",
    "#         sgRNA_rows (list[int] | None)\n",
    "#         font_size (int | str | None)\n",
    "#         clip_on (bool)\n",
    "#         label_at_zero (bool)\n",
    "#         sgRNA_label_sides (list[str] | None): Optional per-sgRNA side for label\n",
    "#             placement. Each entry 'left' or 'right'. If None, all default to\n",
    "#             'left' (existing behavior). Example: ['left','right'] places the\n",
    "#             second label to the right of its rectangle.\n",
    "\n",
    "#     Notes:\n",
    "#         Right-side labels are anchored just beyond the rectangle end (end+1).\n",
    "#     \"\"\"\n",
    "#     if font_size is None:\n",
    "#         font_size = matplotlib.rcParams['font.size']\n",
    "\n",
    "#     if sgRNA_rows is None:\n",
    "#         sgRNA_rows = [0]*len(sgRNA_intervals)\n",
    "#     max_sgRNA_row = max(sgRNA_rows)+1\n",
    "#     this_sgRNA_y_height = sgRNA_y_height/float(max_sgRNA_row)\n",
    "\n",
    "#     # Normalize label sides\n",
    "#     if sgRNA_label_sides is None:\n",
    "#         sgRNA_label_sides = ['left'] * len(sgRNA_intervals)\n",
    "#     else:\n",
    "#         # Pad / truncate to length\n",
    "#         if len(sgRNA_label_sides) < len(sgRNA_intervals):\n",
    "#             sgRNA_label_sides = sgRNA_label_sides + ['left']*(len(sgRNA_intervals)-len(sgRNA_label_sides))\n",
    "#         else:\n",
    "#             sgRNA_label_sides = sgRNA_label_sides[:len(sgRNA_intervals)]\n",
    "\n",
    "#         # Sanitize values\n",
    "#         sgRNA_label_sides = [('right' if s.lower()=='right' else 'left') for s in sgRNA_label_sides]\n",
    "\n",
    "#     min_sgRNA_x = None\n",
    "#     label_left_sgRNA = True\n",
    "\n",
    "#     for idx, sgRNA_int in enumerate(sgRNA_intervals):\n",
    "#         this_sgRNA_start = max(0, sgRNA_int[0])\n",
    "#         this_sgRNA_end = min(sgRNA_int[1], amp_len - 1)\n",
    "#         if this_sgRNA_start > amp_len or this_sgRNA_end < 0:\n",
    "#             continue\n",
    "\n",
    "#         this_sgRNA_y_row_start = sgRNA_y_start + this_sgRNA_y_height*sgRNA_rows[idx]\n",
    "\n",
    "#         # Rectangle\n",
    "#         ax.add_patch(\n",
    "#             patches.Rectangle(\n",
    "#                 (x_offset+this_sgRNA_start, this_sgRNA_y_row_start),\n",
    "#                 1+this_sgRNA_end-this_sgRNA_start,\n",
    "#                 this_sgRNA_y_height,\n",
    "#                 facecolor=(0, 0, 0, 0.15),\n",
    "#                 clip_on=clip_on\n",
    "#             )\n",
    "#         )\n",
    "\n",
    "#         # Clip caps\n",
    "#         if this_sgRNA_start != sgRNA_int[0]:\n",
    "#             ax.add_patch(\n",
    "#                 patches.Rectangle(\n",
    "#                     (x_offset + 0.1 + this_sgRNA_start, this_sgRNA_y_row_start),\n",
    "#                     0.1,\n",
    "#                     this_sgRNA_y_height,\n",
    "#                     facecolor='w',\n",
    "#                     clip_on=clip_on\n",
    "#                 )\n",
    "#             )\n",
    "#         if this_sgRNA_end != sgRNA_int[1]:\n",
    "#             ax.add_patch(\n",
    "#                 patches.Rectangle(\n",
    "#                     (x_offset + 0.8 + this_sgRNA_end, this_sgRNA_y_row_start),\n",
    "#                     0.1,\n",
    "#                     this_sgRNA_y_height,\n",
    "#                     facecolor='w',\n",
    "#                     clip_on=clip_on\n",
    "#                 )\n",
    "#             )\n",
    "\n",
    "#         # Mismatches\n",
    "#         if sgRNA_mismatches is not None and idx < len(sgRNA_mismatches):\n",
    "#             for mismatch in sgRNA_mismatches[idx]:\n",
    "#                 mismatch_plot_pos = sgRNA_int[0] + mismatch\n",
    "#                 if 0 <= mismatch_plot_pos < amp_len:\n",
    "#                     ax.add_patch(\n",
    "#                         patches.Rectangle(\n",
    "#                             (x_offset + mismatch_plot_pos, this_sgRNA_y_row_start),\n",
    "#                             1,\n",
    "#                             this_sgRNA_y_height,\n",
    "#                             facecolor='r',\n",
    "#                             clip_on=clip_on\n",
    "#                         )\n",
    "#                     )\n",
    "\n",
    "#         if min_sgRNA_x is None or this_sgRNA_start < min_sgRNA_x:\n",
    "#             min_sgRNA_x = this_sgRNA_start\n",
    "\n",
    "#         # Label\n",
    "#         if sgRNA_names is not None and idx < len(sgRNA_names) and sgRNA_names[idx] != \"\":\n",
    "#             side = sgRNA_label_sides[idx]\n",
    "#             if side == 'left':\n",
    "#                 # Existing left logic\n",
    "#                 if (label_at_zero and x_offset + this_sgRNA_start < len(sgRNA_names[idx])*0.66):\n",
    "#                     ax.text(\n",
    "#                         0,\n",
    "#                         this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "#                         sgRNA_names[idx] + \" \",\n",
    "#                         horizontalalignment='left',\n",
    "#                         verticalalignment='center',\n",
    "#                         fontsize=font_size\n",
    "#                     )\n",
    "#                 else:\n",
    "#                     ax.text(\n",
    "#                         x_offset + this_sgRNA_start,\n",
    "#                         this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "#                         sgRNA_names[idx] + \" \",\n",
    "#                         horizontalalignment='right',\n",
    "#                         verticalalignment='center',\n",
    "#                         fontsize=font_size\n",
    "#                     )\n",
    "#             else:  # right\n",
    "#                 # Place just after rectangle end\n",
    "#                 label_x = x_offset + this_sgRNA_end + 1.0\n",
    "#                 ax.text(\n",
    "#                     label_x,\n",
    "#                     this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "#                     \" \" + sgRNA_names[idx],\n",
    "#                     horizontalalignment='left',\n",
    "#                     verticalalignment='center',\n",
    "#                     fontsize=font_size\n",
    "#                 )\n",
    "#             label_left_sgRNA = False\n",
    "\n",
    "#     if min_sgRNA_x is not None and label_left_sgRNA:\n",
    "#         if (label_at_zero and x_offset + min_sgRNA_x < 5):\n",
    "#             ax.text(0,\n",
    "#                     this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "#                     'sgRNA ',\n",
    "#                     horizontalalignment='left',\n",
    "#                     verticalalignment='center',\n",
    "#                     fontsize=font_size)\n",
    "#         else:\n",
    "#             ax.text(x_offset+min_sgRNA_x,\n",
    "#                     this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "#                     'sgRNA ',\n",
    "#                     horizontalalignment='right',\n",
    "#                     verticalalignment='center',\n",
    "#                     fontsize=font_size)\n",
    "# // ...existing code...\n",
    "#### End of Modified function to allow right-side labels ####\n",
    "\n",
    "\n",
    "#### Modified function to allow right-side labels and extend grey pegRNA spacerB span ####\n",
    "# ...existing code...\n",
    "def add_sgRNA_to_ax(ax,\n",
    "                    sgRNA_intervals,\n",
    "                    sgRNA_y_start,\n",
    "                    sgRNA_y_height,\n",
    "                    amp_len,\n",
    "                    x_offset=0,\n",
    "                    sgRNA_mismatches=None,\n",
    "                    sgRNA_names=None,\n",
    "                    sgRNA_rows=None,\n",
    "                    font_size=None,\n",
    "                    clip_on=True,\n",
    "                    label_at_zero=False,\n",
    "                    sgRNA_label_sides=None,\n",
    "                    ref_row_seq=None,                 # WT ref (with '-')\n",
    "                    extend_left_non_gap=None):         # per-sgRNA left extension in non-gap bases\n",
    "    \"\"\"\n",
    "    Draw one or more sgRNA annotations on a Matplotlib Axes.\n",
    "\n",
    "    Each sgRNA is a semi-transparent rectangle (start..end). Optional mismatches\n",
    "    are red sub-rectangles. Labels can be placed per sgRNA either to the left\n",
    "    or right of the rectangle via sgRNA_label_sides.\n",
    "\n",
    "    Notes:\n",
    "        - Right-side labels are anchored just beyond the original rectangle end (end+1).\n",
    "        - If extend_left_non_gap is set, the visual rectangle start is shifted\n",
    "          left by that many non-gap bases (skipping '-' in ref_row_seq).\n",
    "        - If ref_row_seq is provided, the rectangle is split into segments and drawn\n",
    "          only over non-gap columns (skipping any '-' within the span).\n",
    "        - Mismatch blocks remain aligned to the original (unextended) sgRNA start.\n",
    "    \"\"\"\n",
    "    if font_size is None:\n",
    "        font_size = matplotlib.rcParams['font.size']\n",
    "\n",
    "    if sgRNA_rows is None:\n",
    "        sgRNA_rows = [0]*len(sgRNA_intervals)\n",
    "    max_sgRNA_row = max(sgRNA_rows)+1\n",
    "    this_sgRNA_y_height = sgRNA_y_height/float(max_sgRNA_row)\n",
    "\n",
    "    # Normalize label sides\n",
    "    if sgRNA_label_sides is None:\n",
    "        sgRNA_label_sides = ['left'] * len(sgRNA_intervals)\n",
    "    else:\n",
    "        if len(sgRNA_label_sides) < len(sgRNA_intervals):\n",
    "            sgRNA_label_sides = sgRNA_label_sides + ['left']*(len(sgRNA_intervals)-len(sgRNA_label_sides))\n",
    "        else:\n",
    "            sgRNA_label_sides = sgRNA_label_sides[:len(sgRNA_intervals)]\n",
    "        sgRNA_label_sides = [('right' if s.lower()=='right' else 'left') for s in sgRNA_label_sides]\n",
    "\n",
    "    # Normalize extension list\n",
    "    if extend_left_non_gap is None:\n",
    "        extend_left_non_gap = [0]*len(sgRNA_intervals)\n",
    "    else:\n",
    "        if len(extend_left_non_gap) < len(sgRNA_intervals):\n",
    "            extend_left_non_gap = extend_left_non_gap + [0]*(len(sgRNA_intervals)-len(extend_left_non_gap))\n",
    "        else:\n",
    "            extend_left_non_gap = extend_left_non_gap[:len(sgRNA_intervals)]\n",
    "\n",
    "    def _left_shift_by_non_gaps(row_seq, start_idx, n_non_gaps):\n",
    "        \"\"\"Return how many columns to shift left to include n_non_gaps non-'-' bases.\"\"\"\n",
    "        if row_seq is None or n_non_gaps <= 0 or start_idx <= 0:\n",
    "            return 0\n",
    "        count = 0\n",
    "        steps = 0\n",
    "        i = int(start_idx) - 1\n",
    "        while i >= 0 and count < n_non_gaps:\n",
    "            if row_seq[i] != '-':\n",
    "                count += 1\n",
    "            steps += 1\n",
    "            i -= 1\n",
    "        return steps\n",
    "\n",
    "    def _non_gap_runs(row_seq, start_idx, end_idx):\n",
    "        \"\"\"Return list of (run_start, run_end) contiguous non-gap segments in [start_idx, end_idx].\"\"\"\n",
    "        if row_seq is None:\n",
    "            return [(int(start_idx), int(end_idx))]\n",
    "        if len(row_seq) == 0:\n",
    "            return []\n",
    "        s = max(0, int(start_idx))\n",
    "        e = min(int(end_idx), len(row_seq) - 1)\n",
    "        runs = []\n",
    "        i = s\n",
    "        while i <= e:\n",
    "            while i <= e and row_seq[i] == '-':\n",
    "                i += 1\n",
    "            if i > e:\n",
    "                break\n",
    "            run_start = i\n",
    "            while i <= e and row_seq[i] != '-':\n",
    "                i += 1\n",
    "            run_end = i - 1\n",
    "            runs.append((run_start, run_end))\n",
    "        return runs\n",
    "\n",
    "    min_sgRNA_x = None\n",
    "    label_left_sgRNA = True\n",
    "\n",
    "    for idx, sgRNA_int in enumerate(sgRNA_intervals):\n",
    "        # Original clipped interval (for mismatch alignment and caps)\n",
    "        this_sgRNA_start = max(0, sgRNA_int[0])\n",
    "        this_sgRNA_end   = min(sgRNA_int[1], amp_len - 1)\n",
    "        if this_sgRNA_start > amp_len or this_sgRNA_end < 0:\n",
    "            continue\n",
    "\n",
    "        this_sgRNA_y_row_start = sgRNA_y_start + this_sgRNA_y_height*sgRNA_rows[idx]\n",
    "\n",
    "        # Visual start extended left by N non-gap bases (skip '-')\n",
    "        left_extra_cols = _left_shift_by_non_gaps(ref_row_seq, this_sgRNA_start, extend_left_non_gap[idx])\n",
    "        display_start = max(0, this_sgRNA_start - left_extra_cols)\n",
    "\n",
    "        # Draw as multiple rectangles over non-gap runs only\n",
    "        runs = _non_gap_runs(ref_row_seq, display_start, this_sgRNA_end)\n",
    "        for seg_start, seg_end in runs:\n",
    "            if seg_start > seg_end:\n",
    "                continue\n",
    "            ax.add_patch(\n",
    "                patches.Rectangle(\n",
    "                    (x_offset + seg_start, this_sgRNA_y_row_start),\n",
    "                    1 + seg_end - seg_start,\n",
    "                    this_sgRNA_y_height,\n",
    "                    facecolor=(0, 0, 0, 0.15),\n",
    "                    clip_on=clip_on\n",
    "                )\n",
    "            )\n",
    "\n",
    "        # Clip caps (based on original interval vs clipping)\n",
    "        if this_sgRNA_start != sgRNA_int[0]:\n",
    "            ax.add_patch(\n",
    "                patches.Rectangle(\n",
    "                    (x_offset + 0.1 + this_sgRNA_start, this_sgRNA_y_row_start),\n",
    "                    0.1,\n",
    "                    this_sgRNA_y_height,\n",
    "                    facecolor='w',\n",
    "                    clip_on=clip_on\n",
    "                )\n",
    "            )\n",
    "        if this_sgRNA_end != sgRNA_int[1]:\n",
    "            ax.add_patch(\n",
    "                patches.Rectangle(\n",
    "                    (x_offset + 0.8 + this_sgRNA_end, this_sgRNA_y_row_start),\n",
    "                    0.1,\n",
    "                    this_sgRNA_y_height,\n",
    "                    facecolor='w',\n",
    "                    clip_on=clip_on\n",
    "                )\n",
    "            )\n",
    "\n",
    "        # Mismatches (relative to original sgRNA start)\n",
    "        if sgRNA_mismatches is not None and idx < len(sgRNA_mismatches):\n",
    "            for mismatch in sgRNA_mismatches[idx]:\n",
    "                mismatch_plot_pos = sgRNA_int[0] + mismatch\n",
    "                if 0 <= mismatch_plot_pos < amp_len:\n",
    "                    ax.add_patch(\n",
    "                        patches.Rectangle(\n",
    "                            (x_offset + mismatch_plot_pos, this_sgRNA_y_row_start),\n",
    "                            1,\n",
    "                            this_sgRNA_y_height,\n",
    "                            facecolor='r',\n",
    "                            clip_on=clip_on\n",
    "                        )\n",
    "                    )\n",
    "\n",
    "        # For left-anchored label and min_x heuristic, use first visible segment start\n",
    "        leftmost_visible = runs[0][0] if runs else display_start\n",
    "        if min_sgRNA_x is None or leftmost_visible < min_sgRNA_x:\n",
    "            min_sgRNA_x = leftmost_visible\n",
    "\n",
    "        # Label\n",
    "        if sgRNA_names is not None and idx < len(sgRNA_names) and sgRNA_names[idx] != \"\":\n",
    "            side = sgRNA_label_sides[idx]\n",
    "            if side == 'left':\n",
    "                anchor_x = x_offset + leftmost_visible\n",
    "                if (label_at_zero and anchor_x < len(sgRNA_names[idx])*0.66):\n",
    "                    ax.text(\n",
    "                        0,\n",
    "                        this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "                        sgRNA_names[idx] + \" \",\n",
    "                        horizontalalignment='left',\n",
    "                        verticalalignment='center',\n",
    "                        fontsize=font_size\n",
    "                    )\n",
    "                else:\n",
    "                    ax.text(\n",
    "                        anchor_x,\n",
    "                        this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "                        sgRNA_names[idx] + \" \",\n",
    "                        horizontalalignment='right',\n",
    "                        verticalalignment='center',\n",
    "                        fontsize=font_size\n",
    "                    )\n",
    "            else:  # right\n",
    "                label_x = x_offset + this_sgRNA_end + 1.0\n",
    "                ax.text(\n",
    "                    label_x,\n",
    "                    this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "                    \" \" + sgRNA_names[idx],\n",
    "                    horizontalalignment='left',\n",
    "                    verticalalignment='center',\n",
    "                    fontsize=font_size\n",
    "                )\n",
    "            label_left_sgRNA = False\n",
    "\n",
    "    if min_sgRNA_x is not None and label_left_sgRNA:\n",
    "        if (label_at_zero and x_offset + min_sgRNA_x < 5):\n",
    "            ax.text(0,\n",
    "                    this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "                    'sgRNA ',\n",
    "                    horizontalalignment='left',\n",
    "                    verticalalignment='center',\n",
    "                    fontsize=font_size)\n",
    "        else:\n",
    "            ax.text(x_offset+min_sgRNA_x,\n",
    "                    this_sgRNA_y_row_start + this_sgRNA_y_height/2,\n",
    "                    'sgRNA ',\n",
    "                    horizontalalignment='right',\n",
    "                    verticalalignment='center',\n",
    "                    fontsize=font_size)\n",
    "# ...existing code...\n",
    "#### End of Modified function to allow right-side labels and extend grey pegRNA spacerB span ####\n",
    "\n",
    "\n",
    "def get_nuc_color(nuc, alpha):\n",
    "    \"\"\"\n",
    "    Return a consistent RGBA color tuple for a nucleotide or special token.\n",
    "\n",
    "    Args:\n",
    "        nuc (str): One of {'A','T','C','G','N','INS','DEL','-'} or any other string.\n",
    "            'N' denotes ambiguous; '-' denotes gap. 'INS'/'DEL' share the same color to\n",
    "            visually group indels. Any unknown token results in a deterministic color\n",
    "            derived from its character codes.\n",
    "        alpha (float): Alpha channel in [0.0, 1.0] controlling transparency.\n",
    "\n",
    "    Returns:\n",
    "        tuple: (r, g, b, a) with values in [0.0, 1.0].\n",
    "    \"\"\"\n",
    "    get_color=lambda x, y, z: (x/255.0, y/255.0, z/255.0, alpha)\n",
    "    if nuc == \"A\":\n",
    "        return get_color(127, 201, 127)\n",
    "    elif nuc == \"T\":\n",
    "        return get_color(190, 174, 212)\n",
    "    elif nuc == \"C\":\n",
    "        return get_color(253, 192, 134)\n",
    "    elif nuc == \"G\":\n",
    "        return get_color(255, 255, 153)\n",
    "    elif nuc == \"N\":\n",
    "        return get_color(200, 200, 200)\n",
    "    elif nuc == \"INS\":\n",
    "#        return get_color(185,219,253)\n",
    "#        return get_color(177,125,76)\n",
    "        return get_color(193, 129, 114)\n",
    "    elif nuc == \"DEL\":\n",
    "        #return get_color(177,125,76)\n",
    "#        return get_color(202,109,87)\n",
    "        return get_color(193, 129, 114)\n",
    "    elif nuc == \"-\":\n",
    "        #return get_color(177,125,76)\n",
    "#        return get_color(202,109,87)\n",
    "        return get_color(30, 30, 30)\n",
    "    elif nuc == \" \":\n",
    "        # white space for padding\n",
    "        return get_color(255, 255, 255)\n",
    "    else: #return a random color (that is based on the nucleotide given)\n",
    "        charSum = 0\n",
    "        for char in nuc.upper():\n",
    "            thisval = ord(char) - 65 #'A' is 65\n",
    "            if thisval < 0 or thisval > 90:\n",
    "                thisval = 0\n",
    "            charSum += thisval\n",
    "        charSum = (charSum/len(nuc))/90.0\n",
    "\n",
    "        return (charSum, (1-charSum), (2*charSum*(1-charSum)))\n",
    "    \n",
    "\n",
    "### Allele plot\n",
    "# We need to customize the seaborn heatmap class and function\n",
    "class Custom_HeatMapper(sns.matrix._HeatMapper):\n",
    "    \"\"\"\n",
    "    Extension of seaborn's private _HeatMapper to support per-element annotation style\n",
    "    (per-element text properties) and to suppress the colorbar.\n",
    "\n",
    "    This utility mirrors seaborn.heatmap internals while allowing an \"annot\" matrix to be\n",
    "    styled cell-by-cell via a parallel matrix of dictionaries (per_element_annot_kws), where\n",
    "    each dictionary can specify matplotlib.text.Text properties for the corresponding cell.\n",
    "\n",
    "    Caution: sns.matrix._HeatMapper is a private API and may change across seaborn versions.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, data, vmin, vmax, cmap, center, robust, annot, fmt,\n",
    "                 annot_kws,per_element_annot_kws,cbar, cbar_kws,\n",
    "                 xticklabels=True, yticklabels=True, mask=None):\n",
    "        \"\"\"\n",
    "        Initialize the heatmap plotter and capture optional per-element annotation styles.\n",
    "\n",
    "        Args mirror seaborn.heatmap/_HeatMapper with the following addition:\n",
    "            per_element_annot_kws (np.ndarray | list | None): Same shape as `annot` where each\n",
    "                element is a dict of Text properties applied to that cell's annotation.\n",
    "                If None, an empty dict is used for every cell.\n",
    "        \"\"\"\n",
    "        super(Custom_HeatMapper, self).__init__(data, vmin, vmax, cmap, center, robust, annot, fmt,\n",
    "                 annot_kws, cbar, cbar_kws,\n",
    "                 xticklabels, yticklabels, mask)\n",
    "\n",
    "        # Prepare a mirror structure for per-element annotation keyword arguments\n",
    "        if annot is not None:\n",
    "            if per_element_annot_kws is None:\n",
    "                self.per_element_annot_kws=np.empty_like(annot, dtype=object)\n",
    "                self.per_element_annot_kws[:]=dict()\n",
    "            else:\n",
    "                self.per_element_annot_kws=per_element_annot_kws\n",
    "\n",
    "    # add per element dict to style the annotation\n",
    "    def _annotate_heatmap(self, ax, mesh):\n",
    "        \"\"\"Add textual labels with the value in each cell.\n",
    "\n",
    "        This override allows passing a per-cell dictionary of Text properties to fine-tune\n",
    "        the appearance (e.g., bold substitutions) while preserving seaborn's luminance-based\n",
    "        foreground color choice.\n",
    "        \"\"\"\n",
    "        mesh.update_scalarmappable()\n",
    "        xpos, ypos = np.meshgrid(ax.get_xticks(), ax.get_yticks())\n",
    "\n",
    "        # Iterate the mesh values, facecolors, annotations, and per-cell styles in lock-step\n",
    "        for x, y, m, color, val, per_element_dict  in zip(xpos.flat, ypos.flat,\n",
    "                                       mesh.get_array().flat, mesh.get_facecolors(),\n",
    "                                       self.annot_data.flat, self.per_element_annot_kws.flat):\n",
    "            #print per_element_dict\n",
    "            if m is not np.ma.masked:\n",
    "                l = sns.utils.relative_luminance(color)\n",
    "                text_color = \".15\" if l > .408 else \"w\"\n",
    "                annotation = (\"{:\" + self.fmt + \"}\").format(str(val))\n",
    "                text_kwargs = dict(color=text_color, ha=\"center\", va=\"center\")\n",
    "                text_kwargs.update(self.annot_kws)\n",
    "                text_kwargs.update(per_element_dict)\n",
    "\n",
    "                ax.text(x, y, annotation, **text_kwargs)\n",
    "\n",
    "\n",
    "    # removed the colorbar\n",
    "    def plot(self, ax, cax, kws):\n",
    "        \"\"\"Draw the heatmap body and tick labels on the provided Axes.\n",
    "\n",
    "        This version deliberately avoids attaching a colorbar and leaves any colorbar\n",
    "        management to the caller.\n",
    "        \"\"\"\n",
    "        # Remove all the Axes spines for a cleaner matrix look\n",
    "        sns.utils.despine(ax=ax, left=True, bottom=True)\n",
    "\n",
    "        # Draw the heatmap as a pcolormesh for performance on large matrices\n",
    "        # If a Normalize is supplied, do NOT also pass vmin/vmax.\n",
    "        if 'norm' in kws and kws['norm'] is not None:\n",
    "            mesh = ax.pcolormesh(self.plot_data, cmap=self.cmap, **kws)\n",
    "        else:\n",
    "            mesh = ax.pcolormesh(self.plot_data, vmin=self.vmin, vmax=self.vmax,\n",
    "                                 cmap=self.cmap, **kws)\n",
    "\n",
    "        # Set axis limits to span the matrix exactly\n",
    "        ax.set(xlim=(0, self.data.shape[1]), ylim=(0, self.data.shape[0]))\n",
    "\n",
    "        # Add row and column labels\n",
    "        ax.set(xticks=self.xticks, yticks=self.yticks)\n",
    "        xtl = ax.set_xticklabels(self.xticklabels)\n",
    "        ytl = ax.set_yticklabels(self.yticklabels, rotation=\"vertical\", va='center')\n",
    "\n",
    "        # Possibly rotate them if they overlap after layout\n",
    "        plt.draw()\n",
    "        if sns.utils.axis_ticklabels_overlap(xtl):\n",
    "            plt.setp(xtl, rotation=\"vertical\")\n",
    "        if sns.utils.axis_ticklabels_overlap(ytl):\n",
    "            plt.setp(ytl, rotation=\"horizontal\")\n",
    "\n",
    "        # Add the axis labels\n",
    "        ax.set(xlabel=self.xlabel, ylabel=self.ylabel)\n",
    "\n",
    "        # Annotate the cells with the formatted values\n",
    "        if self.annot:\n",
    "            self._annotate_heatmap(ax, mesh)\n",
    "\n",
    "\n",
    "def custom_heatmap(data, vmin=None, vmax=None, cmap=None, center=None, robust=False,\n",
    "            annot=None, fmt=\".2g\", annot_kws=None,per_element_annot_kws=None,\n",
    "            linewidths=0, linecolor=\"white\",\n",
    "            cbar=True, cbar_kws=None, cbar_ax=None,\n",
    "            square=False, ax=None, xticklabels=True, yticklabels=True,\n",
    "            mask=None,\n",
    "            **kwargs):\n",
    "    \"\"\"\n",
    "    Convenience wrapper around Custom_HeatMapper to draw a heatmap matrix with optional\n",
    "    per-element annotation styling and without a colorbar by default.\n",
    "\n",
    "    Args:\n",
    "        data (np.ndarray | list): 2D array of numeric values to visualize.\n",
    "        vmin, vmax (float | None): Colormap scaling bounds.\n",
    "        cmap (matplotlib.colors.Colormap | str | None): Colormap to use.\n",
    "        center (float | None): If set, shift the colormap center to this value.\n",
    "        robust (bool): If True, use robust quantiles rather than min/max for colormap scaling.\n",
    "        annot (np.ndarray | list | None): 2D array of values/strings to annotate each cell.\n",
    "        fmt (str): Format string applied to annotations.\n",
    "        annot_kws (dict | None): Global matplotlib.text.Text properties applied to all annotations.\n",
    "        per_element_annot_kws (np.ndarray | list | None): Same shape as annot; each element is a\n",
    "            dict of Text properties applied to that cell, allowing per-cell styles.\n",
    "        linewidths (float): Line width between cells (pcolormesh edge widths).\n",
    "        linecolor (str): Line color between cells.\n",
    "        cbar (bool): Present for API parity; colorbar is not created by this function.\n",
    "        cbar_kws (dict | None): Ignored here; reserved for compatibility.\n",
    "        cbar_ax (matplotlib.axes.Axes | None): Ignored here; reserved for compatibility.\n",
    "        square (bool): If True, set aspect to equal so each cell is square.\n",
    "        ax (matplotlib.axes.Axes | None): Axes to draw into; if None, uses current axes.\n",
    "        xticklabels, yticklabels: Tick label configuration as in seaborn.heatmap.\n",
    "        mask (np.ndarray | None): Boolean mask specifying cells not to plot.\n",
    "        **kwargs: Additional arguments passed to Axes.pcolormesh (e.g., shading, antialiased).\n",
    "\n",
    "    Returns:\n",
    "        matplotlib.axes.Axes: The Axes containing the heatmap.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize the plotter object\n",
    "    plotter = Custom_HeatMapper(data, vmin, vmax, cmap, center, robust, annot, fmt,\n",
    "                          annot_kws, per_element_annot_kws, cbar, cbar_kws, xticklabels,\n",
    "                          yticklabels, mask)\n",
    "\n",
    "\n",
    "    # Add the pcolormesh kwargs here\n",
    "    kwargs[\"linewidths\"] = linewidths\n",
    "    kwargs[\"edgecolor\"] = linecolor\n",
    "\n",
    "    # Draw the plot and return the Axes\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "    if square:\n",
    "        ax.set_aspect(\"equal\")\n",
    "    plotter.plot(ax, cbar_ax, kwargs)\n",
    "    return ax\n",
    "\n",
    "\n",
    "def get_rows_for_sgRNA_annotation(sgRNA_intervals, amp_len):\n",
    "    \"\"\"\n",
    "    Assign a vertical \"row\" for each sgRNA interval so that overlapping intervals\n",
    "    are staggered and do not visually collide when drawn.\n",
    "\n",
    "    The algorithm greedily places each interval on the top-most row that does not\n",
    "    already contain any of its covered x positions. Occupancy is tracked per integer\n",
    "    x position between the interval's (clipped) start and end.\n",
    "\n",
    "    Args:\n",
    "        sgRNA_intervals (list[tuple[int,int]]): List of (start, end) sgRNA spans in reference\n",
    "            coordinates. Intervals are clipped to [0, amp_len-1] for overlap calculations.\n",
    "        amp_len (int): Amplicon/reference length for clipping.\n",
    "\n",
    "    Returns:\n",
    "        np.ndarray: Row indices (int) per sgRNA, where 0 is the top-most row. The rows are\n",
    "        inverted (highest row index becomes 0) so that earlier rows appear visually higher\n",
    "        when drawn relative to negative y offsets.\n",
    "    \"\"\"\n",
    "    # figure out how many rows are needed to show all sgRNAs\n",
    "    sgRNA_plot_rows = [0]*len(sgRNA_intervals) # which row each sgRNA should be plotted on\n",
    "    sgRNA_plot_occupancy = [] # which idxs are already filled on each row\n",
    "    sgRNA_plot_occupancy.append([])\n",
    "    for idx, sgRNA_int in enumerate(sgRNA_intervals):\n",
    "        this_sgRNA_start = max(0, sgRNA_int[0])\n",
    "        this_sgRNA_end = min(sgRNA_int[1], amp_len - 1)\n",
    "        curr_row = 0\n",
    "        if this_sgRNA_start > amp_len or this_sgRNA_end < 0:\n",
    "            # Interval entirely outside; place on row 0 and continue\n",
    "            sgRNA_plot_rows[idx] = curr_row\n",
    "            continue\n",
    "        # Bump the row until there is no position overlap with already-placed intervals\n",
    "        while len(np.intersect1d(sgRNA_plot_occupancy[curr_row], range(this_sgRNA_start, this_sgRNA_end))) > 0:\n",
    "            next_row = curr_row + 1\n",
    "            if not next_row in sgRNA_plot_occupancy:\n",
    "                sgRNA_plot_occupancy.append([])\n",
    "            curr_row = next_row\n",
    "        sgRNA_plot_rows[idx] = curr_row\n",
    "        # Mark occupancy for the chosen row\n",
    "        sgRNA_plot_occupancy[curr_row].extend(range(this_sgRNA_start, this_sgRNA_end))\n",
    "    # Invert rows so that the last-created (lowest) row is drawn lowest when using negative offsets\n",
    "    return(np.subtract(max(sgRNA_plot_rows), sgRNA_plot_rows))\n",
    "\n",
    "\n",
    "# def prep_alleles_table(df_alleles, reference_seq, MAX_N_ROWS, MIN_FREQUENCY):\n",
    "#     \"\"\"\n",
    "#     Prepare matrices and metadata required to render an allele heatmap.\n",
    "\n",
    "#     This function converts a subset of rows from a CRISPResso allele table into:\n",
    "#       - X: numeric matrix encoding per-position allele bases for each displayed allele\n",
    "#       - annot: parallel matrix of string characters for on-cell annotation\n",
    "#       - y_labels: formatted labels per allele row (percent and read count)\n",
    "#       - insertion_dict: mapping of row index to a list of (start,end) x-spans that denote\n",
    "#         insertion events (identified by runs of '-' in the reference alignment)\n",
    "#       - per_element_annot_kws: per-cell Text style dictionaries (e.g., bold substitutions)\n",
    "#       - is_reference: boolean flags indicating whether the allele is identical to the\n",
    "#         provided reference sequence without indels\n",
    "\n",
    "#     Selection is limited to the top MAX_N_ROWS rows that meet MIN_FREQUENCY based on\n",
    "#     df_alleles['%Reads'].\n",
    "\n",
    "#     Args:\n",
    "#         df_alleles (pd.DataFrame): Allele table indexed by Aligned_Sequence (or similar),\n",
    "#             containing at least columns: 'Reference_Sequence', '#Reads', '%Reads'.\n",
    "#         reference_seq (str): Ungapped reference sequence used to determine reference rows.\n",
    "#         MAX_N_ROWS (int): Maximum number of allele rows to include.\n",
    "#         MIN_FREQUENCY (float): Minimum percentage (df_alleles['%Reads']) to include a row.\n",
    "\n",
    "#     Returns:\n",
    "#         tuple:\n",
    "#             X (list[list[int]]): Numeric-encoded alleles using dna_to_numbers mapping.\n",
    "#             annot (list[list[str]]): Same shape as X with literal characters per cell.\n",
    "#             y_labels (list[str]): Display labels per row (e.g., \"12.34% (123 reads)\").\n",
    "#             insertion_dict (defaultdict(list)): row_index -> list of (start, end) insertion spans.\n",
    "#             per_element_annot_kws (np.ndarray): Per-cell dicts of Text properties used for styling.\n",
    "#             is_reference (list[bool]): True if row exactly matches reference_seq and has no indels.\n",
    "#     \"\"\"\n",
    "#     dna_to_numbers={'-':0,'A':1,'T':2,'C':3,'G':4,'N':5}\n",
    "#     seq_to_numbers= lambda seq: [dna_to_numbers[x] for x in seq]\n",
    "#     X=[]\n",
    "#     annot=[]\n",
    "#     y_labels=[]\n",
    "#     insertion_dict=defaultdict(list)\n",
    "#     per_element_annot_kws=[]\n",
    "#     is_reference=[]\n",
    "\n",
    "#     # Regex to find contiguous gap runs ('-') in the reference alignment; these correspond to\n",
    "#     # insertions in the read relative to the reference.\n",
    "#     re_find_indels=re.compile(\"(-*-)\")\n",
    "#     idx_row=0\n",
    "#     for idx, row in df_alleles[df_alleles['%Reads']>=MIN_FREQUENCY][:MAX_N_ROWS].iterrows():\n",
    "#         # Encode the allele (index) sequence\n",
    "#         X.append(seq_to_numbers(idx.upper()))\n",
    "#         annot.append(list(idx))\n",
    "\n",
    "#         # Track insertion spans based on gaps in the reference sequence\n",
    "#         has_indels = False\n",
    "#         for p in re_find_indels.finditer(row['Reference_Sequence']):\n",
    "#             has_indels = True\n",
    "#             insertion_dict[idx_row].append((p.start(), p.end()))\n",
    "\n",
    "#         # Build y-axis labels with percentage and read count\n",
    "#         y_labels.append('%.2f%% (%d reads)' % (row['%Reads'], row['#Reads']))\n",
    "#         if idx == reference_seq and not has_indels:\n",
    "#             is_reference.append(True)\n",
    "#         else:\n",
    "#             is_reference.append(False)\n",
    "\n",
    "#         idx_row+=1\n",
    "\n",
    "#         # Detect substitutions (non-gap mismatches) to style them in bold/black\n",
    "#         idxs_sub= [i_sub for i_sub in range(len(idx)) if \\\n",
    "#                    (row['Reference_Sequence'][i_sub]!=idx[i_sub]) and \\\n",
    "#                    (row['Reference_Sequence'][i_sub]!='-') and\\\n",
    "#                    (idx[i_sub]!='-')]\n",
    "#         to_append=np.array([{}]*len(idx), dtype=object)\n",
    "#         to_append[ idxs_sub]={'weight':'bold', 'color':'black','size':16}\n",
    "#         per_element_annot_kws.append(to_append)\n",
    "\n",
    "#     return X, annot, y_labels, insertion_dict, per_element_annot_kws, is_reference\n",
    "\n",
    "\n",
    "def prep_alleles_table(df_alleles, reference_seq, ref_aln_seq_region, twin_aln_seq_region, MAX_N_ROWS, MIN_FREQUENCY):\n",
    "    \"\"\"\n",
    "    Prepare matrices and metadata required to render an allele heatmap.\n",
    "\n",
    "    This function converts a subset of rows from a CRISPResso allele table into:\n",
    "      - X: numeric matrix encoding per-position allele bases for each displayed allele\n",
    "      - annot: parallel matrix of string characters for on-cell annotation\n",
    "      - y_labels: formatted labels per allele row (percent and read count)\n",
    "      - insertion_dict: mapping of row index to a list of (start,end) x-spans that denote\n",
    "        insertion events (identified by runs of '-' in the reference alignment)\n",
    "      - per_element_annot_kws: per-cell Text style dictionaries (e.g., bold substitutions)\n",
    "      - is_reference: boolean flags indicating whether the allele is identical to the\n",
    "        provided reference sequence without indels\n",
    "\n",
    "    Selection is limited to the top MAX_N_ROWS rows that meet MIN_FREQUENCY based on\n",
    "    df_alleles['%Reads'].\n",
    "\n",
    "    Args:\n",
    "        df_alleles (pd.DataFrame): Allele table indexed by Aligned_Sequence (or similar),\n",
    "            containing at least columns: 'Reference_Sequence', '#Reads', '%Reads'.\n",
    "        reference_seq (str): Ungapped reference sequence used to determine reference rows.\n",
    "        MAX_N_ROWS (int): Maximum number of allele rows to include.\n",
    "        MIN_FREQUENCY (float): Minimum percentage (df_alleles['%Reads']) to include a row.\n",
    "\n",
    "    Returns:\n",
    "        tuple:\n",
    "            X (list[list[int]]): Numeric-encoded alleles using dna_to_numbers mapping.\n",
    "            annot (list[list[str]]): Same shape as X with literal characters per cell.\n",
    "            y_labels (list[str]): Display labels per row (e.g., \"12.34% (123 reads)\").\n",
    "            insertion_dict (defaultdict(list)): row_index -> list of (start, end) insertion spans.\n",
    "            per_element_annot_kws (np.ndarray): Per-cell dicts of Text properties used for styling.\n",
    "            is_reference (list[bool]): True if row exactly matches reference_seq and has no indels.\n",
    "    \"\"\"\n",
    "    dna_to_numbers={'-':0,'A':1,'T':2,'C':3,'G':4,'N':5, ' ':6}\n",
    "    seq_to_numbers= lambda seq: [dna_to_numbers[x] for x in seq]\n",
    "    X=[]\n",
    "    annot=[]\n",
    "    y_labels=[]\n",
    "    insertion_dict=defaultdict(list)\n",
    "    per_element_annot_kws=[]\n",
    "    is_reference=[]\n",
    "    num_blanks = 2\n",
    "\n",
    "    # Regex to find contiguous gap runs ('-') in the reference alignment; these correspond to\n",
    "    # insertions in the read relative to the reference.\n",
    "    re_find_indels=re.compile(\"(-*-)\")\n",
    "    idx_row=0\n",
    "    for idx, row in df_alleles[df_alleles['%Reads']>=MIN_FREQUENCY][:MAX_N_ROWS].iterrows():\n",
    "        # Encode the allele (index) sequence\n",
    "        X.append(seq_to_numbers(idx.upper()))\n",
    "        annot.append(list(idx))\n",
    "\n",
    "        # Track insertion spans based on gaps in the reference sequence\n",
    "        has_indels = False\n",
    "        for p in re_find_indels.finditer(row['Reference_Sequence']):\n",
    "            has_indels = True\n",
    "            insertion_dict[idx_row].append((p.start(), p.end()))\n",
    "\n",
    "        # Build y-axis labels with percentage and read count\n",
    "        y_labels.append('%.2f%% (%d reads)' % (row['%Reads'], row['#Reads']))\n",
    "        if idx == reference_seq and not has_indels:\n",
    "            is_reference.append(True)\n",
    "        else:\n",
    "            is_reference.append(False)\n",
    "\n",
    "        idx_row+=1\n",
    "\n",
    "        # Detect substitutions (non-gap mismatches) to style them in bold/black\n",
    "        idxs_sub= [i_sub for i_sub in range(len(idx)) if \\\n",
    "                   (row['Reference_Sequence'][i_sub]!=idx[i_sub]) and \\\n",
    "                   (row['Reference_Sequence'][i_sub]!='-') and\\\n",
    "                   (idx[i_sub]!='-')]\n",
    "        to_append=np.array([{}]*len(idx), dtype=object)\n",
    "        to_append[ idxs_sub]={'weight':'bold', 'color':'black','size':16}\n",
    "        per_element_annot_kws.append(to_append)\n",
    "\n",
    "    if twin_aln_seq_region is not None and len(twin_aln_seq_region) > 0:\n",
    "        # Append two blank rows to the end of outputs\n",
    "        for _ in range(num_blanks):\n",
    "            blank_row = ' ' * len(twin_aln_seq_region)\n",
    "            X.append(seq_to_numbers(blank_row))\n",
    "            annot.append(list(blank_row))\n",
    "            y_labels.append('')\n",
    "            is_reference.append(False)\n",
    "            to_append=np.array([{'color':'white'}]*len(blank_row), dtype=object)\n",
    "            per_element_annot_kws.append(to_append)\n",
    "\n",
    "        # Append twin_aln_seq_region to the end of outputs\n",
    "        X.append(seq_to_numbers(twin_aln_seq_region.upper()))\n",
    "        annot.append(list(twin_aln_seq_region))\n",
    "        y_labels.append('TwinPE Reference')\n",
    "        is_reference.append(False)\n",
    "        # detect substitutions (non-gap mismatches) between twin_aln_seq_region and ref_aln_seq_region to style them in bold/black\n",
    "        idxs_sub= [i_sub for i_sub in range(len(twin_aln_seq_region)) if \\\n",
    "                   (ref_aln_seq_region[i_sub]!=twin_aln_seq_region[i_sub]) and \\\n",
    "                   (ref_aln_seq_region[i_sub]!='-') and\\\n",
    "                   (twin_aln_seq_region[i_sub]!='-')]\n",
    "        to_append=np.array([{}]*len(twin_aln_seq_region), dtype=object)\n",
    "        to_append[ idxs_sub]={'weight':'bold', 'color':'black','size':16}\n",
    "        per_element_annot_kws.append(to_append)\n",
    "        # track insertions in twin_aln_seq_region relative to ref_aln_seq_region\n",
    "        for p in re_find_indels.finditer(ref_aln_seq_region):\n",
    "            insertion_dict[idx_row+num_blanks].append((p.start(), p.end()))\n",
    "\n",
    "    return X, annot, y_labels, insertion_dict, per_element_annot_kws, is_reference\n",
    "\n",
    "\n",
    "def plot_alleles_heatmap(\n",
    "        reference_seq,\n",
    "        # twin_aln_seq=twin_aln_seq_region,\n",
    "        X,\n",
    "        annot,\n",
    "        y_labels,\n",
    "        insertion_dict,\n",
    "        per_element_annot_kws,\n",
    "        fig_filename_root=None,\n",
    "        custom_colors=None,\n",
    "        SAVE_ALSO_PNG=False,\n",
    "        plot_cut_point=True,\n",
    "        cut_point_ind=None,\n",
    "        sgRNA_intervals=None,\n",
    "        sgRNA_names=None,\n",
    "        sgRNA_mismatches=None,\n",
    "        extend_left_non_gap=None,\n",
    "        category=None,\n",
    "        **kwargs):\n",
    "    \"\"\"\n",
    "    Plots alleles in a heatmap (nucleotides color-coded for easy visualization)\n",
    "    input:\n",
    "    -reference_seq: sequence of reference allele to plot\n",
    "    -X: list of numbers representing nucleotides of the allele\n",
    "    -annot: list of nucleotides (letters) of the allele\n",
    "    -y_labels: list of labels for each row/allele\n",
    "    -insertion_dict: locations of insertions -- red squares will be drawn around these\n",
    "    -per_element_annot_kws: annotations for each cell (e.g. bold for substitutions, etc.)\n",
    "    -fig_filename_root: figure filename to plot (not including '.pdf' or '.png'). If None, plots are shown interactively.\n",
    "    -custom_colors: dict of colors to plot (e.g. colors['A'] = (1,0,0,0.4) # red,blue,green,alpha )\n",
    "    -SAVE_ALSO_PNG: whether to write png file as well\n",
    "    -plot_cut_point: if false, won't draw 'predicted cleavage' line\n",
    "    -cut_point_ind: index of cut point (if None, will be plot in the middle calculated as len(reference_seq)/2)\n",
    "    -sgRNA_intervals: locations where sgRNA is located\n",
    "    -sgRNA_mismatches: array (for each sgRNA_interval) of locations in sgRNA where there are mismatches\n",
    "    -sgRNA_names: array (for each sgRNA_interval) of names of sgRNAs (otherwise empty)\n",
    "    -custom_colors: dict of colors to plot (e.g. colors['A'] = (1,0,0,0.4) # red,blue,green,alpha )\n",
    "    \"\"\"\n",
    "    plot_nuc_len=len(reference_seq)\n",
    "\n",
    "    # make a color map of fixed colors\n",
    "    alpha=0.4\n",
    "    A_color=get_nuc_color('A', alpha)\n",
    "    T_color=get_nuc_color('T', alpha)\n",
    "    C_color=get_nuc_color('C', alpha)\n",
    "    G_color=get_nuc_color('G', alpha)\n",
    "    INDEL_color = get_nuc_color('N', alpha)\n",
    "    blank_color = get_nuc_color(' ', alpha)\n",
    "\n",
    "    if custom_colors is not None:\n",
    "        hex_alpha = '66'  # this is equivalent to 40% in hexadecimal\n",
    "        if 'A' in custom_colors:\n",
    "            A_color = custom_colors['A'] + hex_alpha\n",
    "        if 'T' in custom_colors:\n",
    "            T_color = custom_colors['T'] + hex_alpha\n",
    "        if 'C' in custom_colors:\n",
    "            C_color = custom_colors['C'] + hex_alpha\n",
    "        if 'G' in custom_colors:\n",
    "            G_color = custom_colors['G'] + hex_alpha\n",
    "        if 'N' in custom_colors:\n",
    "            INDEL_color = custom_colors['N'] + hex_alpha\n",
    "        if ' ' in custom_colors:\n",
    "            blank_color = custom_colors[' '] + hex_alpha\n",
    "\n",
    "    dna_to_numbers={'-':0,'A':1,'T':2,'C':3,'G':4,'N':5, ' ':6}\n",
    "    seq_to_numbers= lambda seq: [dna_to_numbers[x] for x in seq]\n",
    "\n",
    "    # cmap = colors_mpl.ListedColormap([INDEL_color, A_color, T_color, C_color, G_color, INDEL_color, blank_color])\n",
    "    # cmap.set_over(blank_color)\n",
    "    # norm = colors_mpl.Normalize(vmin=0, vmax=5)\n",
    "\n",
    "    # New: 7-color colormap + discrete norm for 0..6\n",
    "    cmap = colors_mpl.ListedColormap([INDEL_color, A_color, T_color, C_color, G_color, INDEL_color, blank_color])\n",
    "    bnorm = colors_mpl.BoundaryNorm(np.arange(-0.5, 7.5, 1), cmap.N, clip=False)\n",
    "\n",
    "    #ref_seq_around_cut=reference_seq[max(0,cut_point-plot_nuc_len/2+1):min(len(reference_seq),cut_point+plot_nuc_len/2+1)]\n",
    "\n",
    "#    print('per element anoot kws: ' + per_element_annot_kws)\n",
    "    if len(per_element_annot_kws) > 1:\n",
    "        per_element_annot_kws=np.vstack(per_element_annot_kws[::-1])\n",
    "    else:\n",
    "        per_element_annot_kws=np.array(per_element_annot_kws)\n",
    "    ref_seq_hm=np.expand_dims(seq_to_numbers(reference_seq), 1).T\n",
    "    ref_seq_annot_hm=np.expand_dims(list(reference_seq), 1).T\n",
    "\n",
    "    annot=annot[::-1]\n",
    "    X=X[::-1]\n",
    "\n",
    "    N_ROWS=len(X)\n",
    "    N_COLUMNS=plot_nuc_len\n",
    "\n",
    "    if N_ROWS < 1:\n",
    "        fig, ax = plt.subplots()\n",
    "        fig.text(0.5, 0.5, 'No Alleles', horizontalalignment='center', verticalalignment='center', transform = ax.transAxes)\n",
    "        ax.set_clip_on(False)\n",
    "\n",
    "        if fig_filename_root is None:\n",
    "            plt.show()\n",
    "        else:\n",
    "            fig.savefig(fig_filename_root+'.pdf', bbox_inches='tight')\n",
    "            if SAVE_ALSO_PNG:\n",
    "                fig.savefig(fig_filename_root+'.png', bbox_inches='tight')\n",
    "        plt.close(fig)\n",
    "        return\n",
    "\n",
    "    sgRNA_rows = []\n",
    "    num_sgRNA_rows = 0\n",
    "\n",
    "    if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "        sgRNA_rows = get_rows_for_sgRNA_annotation(sgRNA_intervals, plot_nuc_len)\n",
    "        num_sgRNA_rows = max(sgRNA_rows) + 1\n",
    "        fig=plt.figure(figsize=(plot_nuc_len*0.3, (N_ROWS+1 + num_sgRNA_rows)*0.6))\n",
    "        gs1 = gridspec.GridSpec(N_ROWS+2, N_COLUMNS)\n",
    "        gs2 = gridspec.GridSpec(N_ROWS+2, N_COLUMNS)\n",
    "        #ax_hm_ref heatmap for the reference\n",
    "        ax_hm_ref=plt.subplot(gs1[0:1,:])\n",
    "        ax_hm=plt.subplot(gs2[2:,:])\n",
    "    else:\n",
    "        fig=plt.figure(figsize=(plot_nuc_len*0.3, (N_ROWS+1)*0.6))\n",
    "        gs1 = gridspec.GridSpec(N_ROWS+1, N_COLUMNS)\n",
    "        gs2 = gridspec.GridSpec(N_ROWS+1, N_COLUMNS)\n",
    "        #ax_hm_ref heatmap for the reference\n",
    "        ax_hm_ref=plt.subplot(gs1[0,:])\n",
    "        ax_hm=plt.subplot(gs2[1:,:])\n",
    "\n",
    "\n",
    "    custom_heatmap(ref_seq_hm, annot=ref_seq_annot_hm, annot_kws={'size':16}, cmap=cmap, fmt='s', ax=ax_hm_ref, norm=bnorm, vmin=None, vmax=None, square=True)\n",
    "    custom_heatmap(X, annot=np.array(annot), annot_kws={'size':16}, cmap=cmap, fmt='s', ax=ax_hm, norm=bnorm, vmin=None, vmax=None, square=True, per_element_annot_kws=per_element_annot_kws)\n",
    "    \n",
    "\n",
    "    # place ticks at row centers\n",
    "    ax_hm.yaxis.tick_right()\n",
    "    ax_hm.set_yticks(np.arange(N_ROWS) + 0.5)\n",
    "    # apply labels in reverse order\n",
    "    ax_hm.set_yticklabels(y_labels[::-1], rotation=True, va='center')\n",
    "    # turn off only the 2nd and 3rd tick marks (Python is 0-based)\n",
    "    for idx in [1, 2]:\n",
    "        tick = ax_hm.yaxis.get_major_ticks()[idx]\n",
    "        tick.tick1line.set_visible(False)  # left tick\n",
    "        tick.tick2line.set_visible(False)  # right tick\n",
    "    ax_hm.xaxis.set_ticks([])\n",
    "\n",
    "    #### Orginal method - includes right-side labels####\n",
    "    # if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "    #     this_sgRNA_y_start = -1*num_sgRNA_rows\n",
    "    #     this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "    #     # add_sgRNA_to_ax(ax_hm_ref, sgRNA_intervals, sgRNA_y_start=this_sgRNA_y_start, sgRNA_y_height=this_sgRNA_y_height, amp_len=plot_nuc_len, font_size='small', clip_on=False, sgRNA_names=sgRNA_names, sgRNA_mismatches=sgRNA_mismatches, x_offset=0, label_at_zero=True, sgRNA_rows=sgRNA_rows)\n",
    "    #     add_sgRNA_to_ax(ax_hm_ref, sgRNA_intervals, sgRNA_y_start=this_sgRNA_y_start, sgRNA_y_height=this_sgRNA_y_height, amp_len=plot_nuc_len, font_size='small', clip_on=False, sgRNA_names=sgRNA_names, sgRNA_mismatches=sgRNA_mismatches, x_offset=0, label_at_zero=True, sgRNA_rows=sgRNA_rows, sgRNA_label_sides=['left','right'])\n",
    "    #### Orginal method ####\n",
    "\n",
    "    #### New method - includes right-side labels and extended grey pegRNA spacerB span ####\n",
    "    # if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "    #     this_sgRNA_y_start = -1*num_sgRNA_rows\n",
    "    #     this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "\n",
    "    #     # Extend only the second sgRNA by 3 non-gap bases to the left in WT reference row\n",
    "    #     extend_left = [0]*len(sgRNA_intervals)\n",
    "    #     if len(extend_left) > 1:\n",
    "    #         extend_left[1] = 3\n",
    "\n",
    "    #     add_sgRNA_to_ax(\n",
    "    #         ax_hm_ref,\n",
    "    #         sgRNA_intervals,\n",
    "    #         sgRNA_y_start=this_sgRNA_y_start,\n",
    "    #         sgRNA_y_height=this_sgRNA_y_height,\n",
    "    #         amp_len=plot_nuc_len,\n",
    "    #         font_size='small',\n",
    "    #         clip_on=False,\n",
    "    #         sgRNA_names=sgRNA_names,\n",
    "    #         sgRNA_mismatches=sgRNA_mismatches,\n",
    "    #         x_offset=0,\n",
    "    #         label_at_zero=True,\n",
    "    #         sgRNA_rows=sgRNA_rows,\n",
    "    #         sgRNA_label_sides=['left','right'],\n",
    "    #         ref_row_seq=ref_aln_seq_region,         # WT ref alignment row (with '-')\n",
    "    #         extend_left_non_gap=extend_left         # apply extension\n",
    "    #     )\n",
    "    #### End of New method ####\n",
    "\n",
    "    #### Newer method - includes right-side labels and extended grey pegRNA spacerB span by caller ####\n",
    "    if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "        this_sgRNA_y_start = -1*num_sgRNA_rows\n",
    "        this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "\n",
    "        # Normalize extend_left_non_gap to a list the length of sgRNA_intervals\n",
    "        if extend_left_non_gap is None:\n",
    "            extend = [0]*len(sgRNA_intervals)\n",
    "        elif isinstance(extend_left_non_gap, int):\n",
    "            extend = [extend_left_non_gap]*len(sgRNA_intervals)\n",
    "        elif isinstance(extend_left_non_gap, dict):\n",
    "            # allow mapping by index or (if available) name\n",
    "            extend = [0]*len(sgRNA_intervals)\n",
    "            for i in range(len(sgRNA_intervals)):\n",
    "                extend[i] = extend_left_non_gap.get(\n",
    "                    sgRNA_names[i] if sgRNA_names and i < len(sgRNA_names) else i,\n",
    "                    extend_left_non_gap.get(i, 0)\n",
    "                )\n",
    "        else:\n",
    "            extend = list(extend_left_non_gap)\n",
    "            if len(extend) < len(sgRNA_intervals):\n",
    "                extend += [0]*(len(sgRNA_intervals)-len(extend))\n",
    "            else:\n",
    "                extend = extend[:len(sgRNA_intervals)]\n",
    "\n",
    "        add_sgRNA_to_ax(\n",
    "            ax_hm_ref,\n",
    "            sgRNA_intervals,\n",
    "            sgRNA_y_start=this_sgRNA_y_start,\n",
    "            sgRNA_y_height=this_sgRNA_y_height,\n",
    "            amp_len=plot_nuc_len,\n",
    "            font_size='small',\n",
    "            clip_on=False,\n",
    "            sgRNA_names=sgRNA_names,\n",
    "            sgRNA_mismatches=sgRNA_mismatches,\n",
    "            x_offset=0,\n",
    "            label_at_zero=True,\n",
    "            sgRNA_rows=sgRNA_rows,\n",
    "            sgRNA_label_sides=['left','right'],\n",
    "            ref_row_seq=ref_aln_seq_region,     # WT ref alignment row (with '-')\n",
    "            extend_left_non_gap=extend          # apply caller-provided extension\n",
    "        )\n",
    "    #### End of Newer method ####\n",
    "\n",
    "    #create boxes for ins\n",
    "    for idx, lss in insertion_dict.items():\n",
    "        for ls in lss:\n",
    "            ax_hm.add_patch(patches.Rectangle((ls[0], N_ROWS-idx-1), ls[1]-ls[0], 1, linewidth=3, edgecolor='r', fill=False))\n",
    "\n",
    "    #### Original method\n",
    "    # #cut point vertical line\n",
    "    # if plot_cut_point:\n",
    "    #     if cut_point_ind is None:\n",
    "    #         cut_point_ind = [plot_nuc_len / 2]\n",
    "    #     # ax_hm.vlines(cut_point_ind,*ax_hm.get_ylim(),linestyles='dashed')\n",
    "    #     ax_hm.vlines(180, 3, N_ROWS, linestyles='dashed')\n",
    "\n",
    "    #### New method\n",
    "    #cut point vertical line\n",
    "    # if plot_cut_point:\n",
    "    #     positions = cut_point_ind\n",
    "    #     if positions is None:\n",
    "    #         positions = [plot_nuc_len / 2]\n",
    "    #     elif not isinstance(positions, (list, tuple, np.ndarray)):\n",
    "    #         positions = [positions]\n",
    "    #     ymin, ymax = ax_hm.get_ylim()\n",
    "    #     for pos in positions:\n",
    "    #         ax_hm.vlines(pos, ymin, ymax, linestyles='dashed')\n",
    "    #         print(pos)\n",
    "\n",
    "#### Newer method to draw cut point lines at correct position and for WT and TwinPE refs rows ####\n",
    "    # cut point vertical line\n",
    "    if plot_cut_point:\n",
    "        if cut_point_ind is None:\n",
    "            cut_point_ind = [plot_nuc_len / 2]\n",
    "\n",
    "        def _is_blank_row(chars):\n",
    "            return all(c == \" \" for c in chars)\n",
    "\n",
    "        N_ROWS = len(annot)\n",
    "\n",
    "        n_bottom_blank = 0\n",
    "        for i in range(1, min(3, N_ROWS)):\n",
    "            if _is_blank_row(annot[i]):\n",
    "                n_bottom_blank += 1\n",
    "            else:\n",
    "                break\n",
    "\n",
    "        # Convert base indices to boundary positions (draw line AFTER the specified base).\n",
    "        # If cut_point_ind values are 0-based indices of the cut base, we add +1 so the\n",
    "        # vertical line appears between base i and i+1 (user expectation).\n",
    "        raw_points = cut_point_ind if isinstance(cut_point_ind, (list, tuple, np.ndarray)) else [cut_point_ind]\n",
    "        xs = [cp + 1 for cp in raw_points if cp is not None]\n",
    "\n",
    "        # On allele table (ax_hm): draw over TwinPE row\n",
    "        if N_ROWS >= 1:\n",
    "            ax_hm.vlines(xs, 0, 1, linestyles=\"dashed\")\n",
    "\n",
    "        # Skip the blank spacer rows and draw through remaining allele rows\n",
    "        y_start = 1 + n_bottom_blank\n",
    "        if N_ROWS > y_start:\n",
    "            ax_hm.vlines(xs, y_start, N_ROWS, linestyles=\"dashed\")\n",
    "\n",
    "        # On WT Reference (ax_hm_ref): single row\n",
    "        ax_hm_ref.vlines(xs, 0, 1, linestyles=\"dashed\")\n",
    "    #### End of New method ####\n",
    "    ax_hm_ref.yaxis.tick_right()\n",
    "    ax_hm_ref.xaxis.set_ticks([])\n",
    "    ax_hm_ref.yaxis.set_ticklabels(['WT Reference'], rotation=True, va='center')\n",
    "\n",
    "    gs2.update(left=0, right=1, hspace=0.05, wspace=0, top=1*(((N_ROWS)*1.13))/(N_ROWS))\n",
    "    gs1.update(left=0, right=1, hspace=0.05, wspace=0,)\n",
    "\n",
    "    sns.set_context(rc={'axes.facecolor':'white','lines.markeredgewidth': 1,'mathtext.fontset' : 'stix','text.usetex':True,'text.latex.unicode':True} )\n",
    "\n",
    "    proxies = [matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='black',\n",
    "                    mec='none', marker=r'$\\mathbf{{{}}}$'.format('bold'), ms=18),\n",
    "               matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "                    mec='r', marker='s', ms=8, markeredgewidth=2.5),\n",
    "              matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "                    mec='black', marker='_', ms=2,)]\n",
    "    descriptions=['Substitutions', 'Insertions', 'Deletions']\n",
    "\n",
    "    if plot_cut_point:\n",
    "        proxies.append(\n",
    "              matplotlib.lines.Line2D([0], [1], linestyle='--', c='black', ms=6))\n",
    "        descriptions.append('Predicted cleavage position')\n",
    "\n",
    "    if category:\n",
    "        category = category.replace(\"_\", r\"\\ \")\n",
    "        proxies.append(patches.Patch(color='none'))\n",
    "        descriptions.append(rf\"$\\bf{{{category}\\ Alleles}}$\")\n",
    "\n",
    "    #ax_hm_ref.legend(proxies, descriptions, numpoints=1, markerscale=2, loc='center', bbox_to_anchor=(0.5, 4),ncol=1)\n",
    "    lgd = ax_hm.legend(proxies, descriptions, numpoints=1, markerscale=2, loc='upper center', bbox_to_anchor=(0.5, 0), ncol=1, fancybox=True, shadow=False)\n",
    "\n",
    "    if fig_filename_root is None:\n",
    "        plt.show()\n",
    "    else:\n",
    "        fig.savefig(fig_filename_root+'.pdf', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "        if SAVE_ALSO_PNG:\n",
    "            fig.savefig(fig_filename_root+'.png', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "    plt.close(fig)\n",
    "\n",
    "\n",
    "# def plot_alleles_heatmap(\n",
    "#         reference_seq,\n",
    "#         twin_aln_seq,\n",
    "#         X,\n",
    "#         annot,\n",
    "#         y_labels,\n",
    "#         insertion_dict,\n",
    "#         per_element_annot_kws,\n",
    "#         fig_filename_root=None,\n",
    "#         custom_colors=None,\n",
    "#         SAVE_ALSO_PNG=False,\n",
    "#         plot_cut_point=True,\n",
    "#         cut_point_ind=None,\n",
    "#         sgRNA_intervals=None,\n",
    "#         sgRNA_names=None,\n",
    "#         sgRNA_mismatches=None,\n",
    "#         category=None,\n",
    "#         **kwargs):\n",
    "#     \"\"\"\n",
    "#     Render a nucleotide-annotated heatmap of allele sequences with an optional reference\n",
    "#     row, optional sgRNA overlays, and optional cleavage position line.\n",
    "\n",
    "#     Args:\n",
    "#         reference_seq (str): Ungapped reference sequence shown as the top (reference) row.\n",
    "#         X (list[list[int]] | np.ndarray): Matrix of numeric-encoded nucleotides per allele row.\n",
    "#         annot (list[list[str]] | np.ndarray): Matrix of characters for per-cell annotation\n",
    "#             (same shape/order as X).\n",
    "#         y_labels (list[str]): Labels for each allele row in X (rendered along the y-axis).\n",
    "#         insertion_dict (dict[int, list[tuple[int,int]]]): For each row index, spans to outline\n",
    "#             as insertions (drawn as red boxes around the cells covering the span).\n",
    "#         per_element_annot_kws (np.ndarray): Per-cell dicts of Text properties for annotation styling.\n",
    "#         fig_filename_root (str | None): If None, display interactively; else save as <root>.pdf\n",
    "#             (and optionally <root>.png when SAVE_ALSO_PNG is True).\n",
    "#         custom_colors (dict | None): Optional mapping like {'A': '#hex', 'T': '#hex', 'C': '#hex',\n",
    "#             'G': '#hex', 'N': '#hex'}; alpha is appended automatically (~40%).\n",
    "#         SAVE_ALSO_PNG (bool): If True and fig_filename_root is set, also write PNG alongside PDF.\n",
    "#         plot_cut_point (bool): If True, draw a vertical dashed line at cut_point_ind (or center).\n",
    "#         cut_point_ind (int | list[int] | None): One or more x indices for predicted cleavage.\n",
    "#             If None, defaults to the midpoint of the reference length.\n",
    "#         sgRNA_intervals (list[tuple[int,int]] | None): Optional sgRNA spans to overlay above the\n",
    "#             reference row.\n",
    "#         sgRNA_names (list[str] | None): Optional per-sgRNA labels.\n",
    "#         sgRNA_mismatches (list[list[int]] | None): Optional per-sgRNA mismatch positions (relative\n",
    "#             to the sgRNA start) to draw in red.\n",
    "#         **kwargs: Forwarded to the underlying pcolormesh via custom_heatmap (e.g., shading).\n",
    "\n",
    "#     Returns:\n",
    "#         None. Displays or saves the figure and closes it.\n",
    "#     \"\"\"\n",
    "#     plot_nuc_len=len(reference_seq)\n",
    "\n",
    "#     # make a color map of fixed colors\n",
    "#     alpha=0.4\n",
    "#     A_color=get_nuc_color('A', alpha)\n",
    "#     T_color=get_nuc_color('T', alpha)\n",
    "#     C_color=get_nuc_color('C', alpha)\n",
    "#     G_color=get_nuc_color('G', alpha)\n",
    "#     INDEL_color = get_nuc_color('N', alpha)\n",
    "\n",
    "#     # Allow overriding base colors with hex strings; append hex alpha for ~40% transparency\n",
    "#     if custom_colors is not None:\n",
    "#         hex_alpha = '66'  # this is equivalent to 40% in hexadecimal\n",
    "#         if 'A' in custom_colors:\n",
    "#             A_color = custom_colors['A'] + hex_alpha\n",
    "#         if 'T' in custom_colors:\n",
    "#             T_color = custom_colors['T'] + hex_alpha\n",
    "#         if 'C' in custom_colors:\n",
    "#             C_color = custom_colors['C'] + hex_alpha\n",
    "#         if 'G' in custom_colors:\n",
    "#             G_color = custom_colors['G'] + hex_alpha\n",
    "#         if 'N' in custom_colors:\n",
    "#             INDEL_color = custom_colors['N'] + hex_alpha\n",
    "\n",
    "#     dna_to_numbers={'-':0,'A':1,'T':2,'C':3,'G':4,'N':5}\n",
    "#     seq_to_numbers= lambda seq: [dna_to_numbers[x] for x in seq]\n",
    "\n",
    "#     # Build the ListedColormap using the ordered mapping; the last 'N' duplicates the first to\n",
    "#     # ensure gaps/ambiguous map consistently.\n",
    "#     cmap = colors_mpl.ListedColormap([INDEL_color, A_color, T_color, C_color, G_color, INDEL_color])\n",
    "\n",
    "#     #ref_seq_around_cut=reference_seq[max(0,cut_point-plot_nuc_len/2+1):min(len(reference_seq),cut_point+plot_nuc_len/2+1)]\n",
    "\n",
    "# #    print('per element anoot kws: ' + per_element_annot_kws)\n",
    "#     if len(per_element_annot_kws) > 1:\n",
    "#         # Stack per-row annotation styles and invert order to align with inverted X below\n",
    "#         per_element_annot_kws=np.vstack(per_element_annot_kws[::-1])\n",
    "#     else:\n",
    "#         per_element_annot_kws=np.array(per_element_annot_kws)\n",
    "#     ref_seq_hm=np.expand_dims(seq_to_numbers(reference_seq), 1).T\n",
    "#     ref_seq_annot_hm=np.expand_dims(list(reference_seq), 1).T\n",
    "#     twin_seq_hm=np.expand_dims(seq_to_numbers(twin_aln_seq), 1).T\n",
    "#     twin_seq_annot_hm=np.expand_dims(list(twin_aln_seq), 1).T\n",
    "\n",
    "#     # Reverse to plot the most frequent allele at the top of the heatmap\n",
    "#     annot=annot[::-1]\n",
    "#     X=X[::-1]\n",
    "\n",
    "#     N_ROWS=len(X)\n",
    "#     N_COLUMNS=plot_nuc_len\n",
    "\n",
    "#     if N_ROWS < 1:\n",
    "#         # No alleles to show; emit a placeholder figure\n",
    "#         fig, ax = plt.subplots()\n",
    "#         fig.text(0.5, 0.5, 'No Alleles', horizontalalignment='center', verticalalignment='center', transform = ax.transAxes)\n",
    "#         ax.set_clip_on(False)\n",
    "\n",
    "#         if fig_filename_root is None:\n",
    "#             plt.show()\n",
    "#         else:\n",
    "#             fig.savefig(fig_filename_root+'.pdf', bbox_inches='tight')\n",
    "#             if SAVE_ALSO_PNG:\n",
    "#                 fig.savefig(fig_filename_root+'.png', bbox_inches='tight')\n",
    "#         plt.close(fig)\n",
    "#         return\n",
    "\n",
    "# # ...existing code...\n",
    "#     # Determine how much vertical space is needed to accommodate optional sgRNA rows\n",
    "#     sgRNA_rows = []\n",
    "#     num_sgRNA_rows = 0\n",
    "#     if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "#         sgRNA_rows = get_rows_for_sgRNA_annotation(sgRNA_intervals, plot_nuc_len)\n",
    "#         num_sgRNA_rows = max(sgRNA_rows) + 1\n",
    "\n",
    "#     # Size the figure to satisfy square cells for the allele heatmap\n",
    "#     cells_per_inch = 0.30  # horizontal size per base (tweak to taste)\n",
    "#     fig_w = plot_nuc_len * cells_per_inch\n",
    "\n",
    "#     # Give the ref axis extra room for sgRNA rows\n",
    "#     ref_rows   = 1 + num_sgRNA_rows\n",
    "#     hm_rows    = N_ROWS\n",
    "#     twin_rows  = 1\n",
    "#     total_rows = ref_rows + hm_rows + twin_rows\n",
    "\n",
    "#     # Height to keep the central heatmap square: H = W * (total_rows / N_COLUMNS)\n",
    "#     fig_h = fig_w * (total_rows / float(N_COLUMNS))\n",
    "\n",
    "#     # Single GridSpec with proportional heights\n",
    "#     fig = plt.figure(figsize=(fig_w, fig_h), constrained_layout=True)\n",
    "#     gs = gridspec.GridSpec(3, N_COLUMNS, height_ratios=[ref_rows, hm_rows, twin_rows], figure=fig)\n",
    "\n",
    "#     ax_hm_ref  = fig.add_subplot(gs[0, :])\n",
    "#     ax_hm      = fig.add_subplot(gs[1, :])\n",
    "#     ax_hm_twin = fig.add_subplot(gs[2, :])\n",
    "\n",
    "#     # Draw reference row (top), allele matrix (middle), twin row (bottom)\n",
    "#     custom_heatmap(ref_seq_hm,  annot=ref_seq_annot_hm,  annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm_ref,  vmin=0, vmax=5, square=False)\n",
    "#     custom_heatmap(X,           annot=np.array(annot),   annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm,     vmin=0, vmax=5, square=True, per_element_annot_kws=per_element_annot_kws)\n",
    "#     custom_heatmap(twin_seq_hm, annot=twin_seq_annot_hm, annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm_twin, vmin=0, vmax=5, square=False)\n",
    "\n",
    "#     # Expand ref ylim to show sgRNA rows inside the ref axes\n",
    "#     if num_sgRNA_rows > 0:\n",
    "#         ax_hm_ref.set_ylim(-num_sgRNA_rows, 1)\n",
    "\n",
    "#     # Remove previous aspect/anchor/layout hacks\n",
    "#     # if hasattr(ax_hm, \"set_box_aspect\"): ...\n",
    "#     # gs.update(...)\n",
    "\n",
    "#     # Y on the right to keep allele labels near the matrix; hide x ticks (letters are the labels)\n",
    "#     ax_hm.yaxis.tick_right()\n",
    "#     ax_hm.yaxis.set_ticklabels(y_labels[::-1], rotation=True, va='center')\n",
    "#     ax_hm.xaxis.set_ticks([])\n",
    "\n",
    "#     # Draw sgRNA overlays above the ref row (now there is real space in the axes)\n",
    "#     if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "#         this_sgRNA_y_start = -1*num_sgRNA_rows\n",
    "#         this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "#         add_sgRNA_to_ax(ax_hm_ref, sgRNA_intervals,\n",
    "#                         sgRNA_y_start=this_sgRNA_y_start,\n",
    "#                         sgRNA_y_height=this_sgRNA_y_height,\n",
    "#                         amp_len=plot_nuc_len,\n",
    "#                         font_size='small',\n",
    "#                         clip_on=True,  # now safe to clip because ylim was expanded\n",
    "#                         sgRNA_names=sgRNA_names,\n",
    "#                         sgRNA_mismatches=sgRNA_mismatches,\n",
    "#                         x_offset=0,\n",
    "#                         label_at_zero=True,\n",
    "#                         sgRNA_rows=sgRNA_rows)\n",
    "\n",
    "#     # Reference axis cosmetics\n",
    "#     ax_hm_ref.yaxis.tick_right()\n",
    "#     ax_hm_ref.xaxis.set_ticks([])\n",
    "#     ax_hm_ref.yaxis.set_ticklabels(['WT Reference'], rotation=True, va='center')\n",
    "\n",
    "#     # Twin seq axis cosmetics\n",
    "#     ax_hm_twin.yaxis.tick_right()\n",
    "#     ax_hm_twin.xaxis.set_ticks([])\n",
    "#     ax_hm_twin.yaxis.set_ticklabels(['TwinPE Reference'], rotation=True, va='center')\n",
    "# # ...existing code...\n",
    "\n",
    "#     # Draw sgRNA overlays just above the reference row when provided\n",
    "#     if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "#         this_sgRNA_y_start = -1*num_sgRNA_rows\n",
    "#         this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "#         add_sgRNA_to_ax(ax_hm_ref, sgRNA_intervals, sgRNA_y_start=this_sgRNA_y_start, sgRNA_y_height=this_sgRNA_y_height, amp_len=plot_nuc_len, font_size='small', clip_on=False, sgRNA_names=sgRNA_names, sgRNA_mismatches=sgRNA_mismatches, x_offset=0, label_at_zero=True, sgRNA_rows=sgRNA_rows)\n",
    "\n",
    "#     # Outline insertion spans in red for each allele row\n",
    "#     for idx, lss in insertion_dict.items():\n",
    "#         for ls in lss:\n",
    "#             ax_hm.add_patch(patches.Rectangle((ls[0], N_ROWS-idx-1), ls[1]-ls[0], 1, linewidth=3, edgecolor='r', fill=False))\n",
    "\n",
    "#     # Draw predicted cleavage site(s) as vertical dashed line(s)\n",
    "#     if plot_cut_point:\n",
    "#         if cut_point_ind is None:\n",
    "#             cut_point_ind = [plot_nuc_len / 2]\n",
    "#         ax_hm.vlines(cut_point_ind,*ax_hm.get_ylim(),linestyles='dashed')\n",
    "\n",
    "#     # Reference axis cosmetics\n",
    "#     ax_hm_ref.yaxis.tick_right()\n",
    "#     ax_hm_ref.xaxis.set_ticks([])\n",
    "#     ax_hm_ref.yaxis.set_ticklabels(['WT Reference'], rotation=True, va='center')\n",
    "    \n",
    "#     # Twin seq axis cosmetics\n",
    "#     ax_hm_twin.yaxis.tick_right()\n",
    "#     ax_hm_twin.xaxis.set_ticks([])\n",
    "#     ax_hm_twin.yaxis.set_ticklabels(['TwinPE Reference'], rotation=True, va='center')\n",
    "\n",
    "#     # Tighten spacing; keep top margin roughly constant across different N_ROWS\n",
    "#     # gs2.update(left=0, right=1, hspace=0.05, wspace=0, top=1*(((N_ROWS)*1.13))/(N_ROWS))\n",
    "#     # gs1.update(left=0, right=1, hspace=0.05, wspace=0,)\n",
    "#     # gs3.update(left=0, right=1, hspace=0.05, wspace=0,)\n",
    "\n",
    "#     # Styling context for consistent fonts and math text\n",
    "#     sns.set_context(rc={'axes.facecolor':'white','lines.markeredgewidth': 1,'mathtext.fontset' : 'stix','text.usetex':True,'text.latex.unicode':True} )\n",
    "\n",
    "#     # Legend proxies to communicate substitution/indel encodings\n",
    "#     proxies = [matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='black',\n",
    "#                     mec='none', marker=r'$\\mathbf{{{}}}$'.format('bold'), ms=18),\n",
    "#                matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "#                     mec='r', marker='s', ms=8, markeredgewidth=2.5),\n",
    "#               matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "#                     mec='black', marker='_', ms=2,)]\n",
    "#     descriptions=['Substitutions', 'Insertions', 'Deletions']\n",
    "\n",
    "#     if plot_cut_point:\n",
    "#         proxies.append(\n",
    "#               matplotlib.lines.Line2D([0], [1], linestyle='--', c='black', ms=6))\n",
    "#         descriptions.append('Predicted cleavage position')\n",
    "\n",
    "#     if category:\n",
    "#         category = category.replace(\"_\", r\"\\ \")\n",
    "#         proxies.append(patches.Patch(color='none'))\n",
    "#         descriptions.append(rf\"$\\bf{{{category}\\ Alleles}}$\")\n",
    "\n",
    "\n",
    "#     # Place legend under the twin seq heatmap\n",
    "#     #ax_hm_ref.legend(proxies, descriptions, numpoints=1, markerscale=2, loc='center', bbox_to_anchor=(0.5, 4),ncol=1)\n",
    "#     lgd = ax_hm_twin.legend(proxies, descriptions, numpoints=1, markerscale=2, loc='upper center', bbox_to_anchor=(0.5, 0), ncol=1, fancybox=True, shadow=False)\n",
    "\n",
    "#     # # Place title above the entire figure\n",
    "#     # if category:\n",
    "#     #     fig.suptitle(f\"{category} Alleles\", fontsize=24, y=0.5, x=0.25)\n",
    "\n",
    "#     # Display or save\n",
    "#     if fig_filename_root is None:\n",
    "#         plt.show()\n",
    "#     else:\n",
    "#         fig.savefig(fig_filename_root+'.pdf', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "#         if SAVE_ALSO_PNG:\n",
    "#             fig.savefig(fig_filename_root+'.png', bbox_inches='tight', bbox_extra_artists=(lgd,))\n",
    "#     plt.close(fig)\n",
    "\n",
    "# def plot_alleles_heatmap(\n",
    "#         reference_seq,\n",
    "#         twin_aln_seq,\n",
    "#         X,\n",
    "#         annot,\n",
    "#         y_labels,\n",
    "#         insertion_dict,\n",
    "#         per_element_annot_kws,\n",
    "#         fig_filename_root=None,\n",
    "#         custom_colors=None,\n",
    "#         SAVE_ALSO_PNG=False,\n",
    "#         plot_cut_point=True,\n",
    "#         cut_point_ind=None,\n",
    "#         sgRNA_intervals=None,\n",
    "#         sgRNA_names=None,\n",
    "#         sgRNA_mismatches=None,\n",
    "#         category=None,\n",
    "#         **kwargs):\n",
    "#     \"\"\"\n",
    "#     Render a nucleotide-annotated heatmap of allele sequences with an optional reference\n",
    "#     row, optional sgRNA overlays, and optional cleavage position line.\n",
    "\n",
    "#     Args:\n",
    "#         reference_seq (str): Ungapped reference sequence shown as the top (reference) row.\n",
    "#         X (list[list[int]] | np.ndarray): Matrix of numeric-encoded nucleotides per allele row.\n",
    "#         annot (list[list[str]] | np.ndarray): Matrix of characters for per-cell annotation\n",
    "#             (same shape/order as X).\n",
    "#         y_labels (list[str]): Labels for each allele row in X (rendered along the y-axis).\n",
    "#         insertion_dict (dict[int, list[tuple[int,int]]]): For each row index, spans to outline\n",
    "#             as insertions (drawn as red boxes around the cells covering the span).\n",
    "#         per_element_annot_kws (np.ndarray): Per-cell dicts of Text properties for annotation styling.\n",
    "#         fig_filename_root (str | None): If None, display interactively; else save as <root>.pdf\n",
    "#             (and optionally <root>.png when SAVE_ALSO_PNG is True).\n",
    "#         custom_colors (dict | None): Optional mapping like {'A': '#hex', 'T': '#hex', 'C': '#hex',\n",
    "#             'G': '#hex', 'N': '#hex'}; alpha is appended automatically (~40%).\n",
    "#         SAVE_ALSO_PNG (bool): If True and fig_filename_root is set, also write PNG alongside PDF.\n",
    "#         plot_cut_point (bool): If True, draw a vertical dashed line at cut_point_ind (or center).\n",
    "#         cut_point_ind (int | list[int] | None): One or more x indices for predicted cleavage.\n",
    "#             If None, defaults to the midpoint of the reference length.\n",
    "#         sgRNA_intervals (list[tuple[int,int]] | None): Optional sgRNA spans to overlay above the\n",
    "#             reference row.\n",
    "#         sgRNA_names (list[str] | None): Optional per-sgRNA labels.\n",
    "#         sgRNA_mismatches (list[list[int]] | None): Optional per-sgRNA mismatch positions (relative\n",
    "#             to the sgRNA start) to draw in red.\n",
    "#         **kwargs: Forwarded to the underlying pcolormesh via custom_heatmap (e.g., shading).\n",
    "\n",
    "#     Returns:\n",
    "#         None. Displays or saves the figure and closes it.\n",
    "#     \"\"\"\n",
    "#     plot_nuc_len = len(reference_seq)\n",
    "\n",
    "#     # make a color map of fixed colors\n",
    "#     alpha = 0.4\n",
    "#     A_color = get_nuc_color('A', alpha)\n",
    "#     T_color = get_nuc_color('T', alpha)\n",
    "#     C_color = get_nuc_color('C', alpha)\n",
    "#     G_color = get_nuc_color('G', alpha)\n",
    "#     INDEL_color = get_nuc_color('N', alpha)\n",
    "\n",
    "#     # Allow overriding base colors with hex strings; append hex alpha for ~40% transparency\n",
    "#     if custom_colors is not None:\n",
    "#         hex_alpha = '66'  # this is equivalent to 40% in hexadecimal\n",
    "#         if 'A' in custom_colors:\n",
    "#             A_color = custom_colors['A'] + hex_alpha\n",
    "#         if 'T' in custom_colors:\n",
    "#             T_color = custom_colors['T'] + hex_alpha\n",
    "#         if 'C' in custom_colors:\n",
    "#             C_color = custom_colors['C'] + hex_alpha\n",
    "#         if 'G' in custom_colors:\n",
    "#             G_color = custom_colors['G'] + hex_alpha\n",
    "#         if 'N' in custom_colors:\n",
    "#             INDEL_color = custom_colors['N'] + hex_alpha\n",
    "\n",
    "#     dna_to_numbers = {'-':0,'A':1,'T':2,'C':3,'G':4,'N':5}\n",
    "#     seq_to_numbers = lambda seq: [dna_to_numbers[x] for x in seq]\n",
    "\n",
    "#     # Build the ListedColormap using the ordered mapping; the last 'N' duplicates the first to\n",
    "#     # ensure gaps/ambiguous map consistently.\n",
    "#     cmap = colors_mpl.ListedColormap([INDEL_color, A_color, T_color, C_color, G_color, INDEL_color])\n",
    "\n",
    "#     # Prepare per-element styles\n",
    "#     if len(per_element_annot_kws) > 1:\n",
    "#         per_element_annot_kws = np.vstack(per_element_annot_kws[::-1])\n",
    "#     else:\n",
    "#         per_element_annot_kws = np.array(per_element_annot_kws)\n",
    "\n",
    "#     # Heatmap matrices for reference/twin (1 x N) and alleles (N_ROWS x N_COLUMNS)\n",
    "#     ref_seq_hm        = np.expand_dims(seq_to_numbers(reference_seq), 1).T\n",
    "#     ref_seq_annot_hm  = np.expand_dims(list(reference_seq), 1).T\n",
    "#     twin_seq_hm       = np.expand_dims(seq_to_numbers(twin_aln_seq), 1).T\n",
    "#     twin_seq_annot_hm = np.expand_dims(list(twin_aln_seq), 1).T\n",
    "\n",
    "#     # Reverse to plot the most frequent allele at the top of the heatmap\n",
    "#     annot = annot[::-1]\n",
    "#     X     = X[::-1]\n",
    "\n",
    "#     N_ROWS    = len(X)\n",
    "#     N_COLUMNS = plot_nuc_len\n",
    "\n",
    "#     if N_ROWS < 1:\n",
    "#         # No alleles to show; emit a placeholder figure\n",
    "#         fig, ax = plt.subplots()\n",
    "#         fig.text(0.5, 0.5, 'No Alleles', horizontalalignment='center', verticalalignment='center', transform=ax.transAxes)\n",
    "#         ax.set_clip_on(False)\n",
    "\n",
    "#         if fig_filename_root is None:\n",
    "#             plt.show()\n",
    "#         else:\n",
    "#             fig.savefig(fig_filename_root + '.pdf')\n",
    "#             if SAVE_ALSO_PNG:\n",
    "#                 fig.savefig(fig_filename_root + '.png')\n",
    "#         plt.close(fig)\n",
    "#         return\n",
    "\n",
    "#     # Determine how much vertical space is needed to accommodate optional sgRNA rows\n",
    "#     sgRNA_rows = []\n",
    "#     num_sgRNA_rows = 0\n",
    "#     if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "#         sgRNA_rows = get_rows_for_sgRNA_annotation(sgRNA_intervals, plot_nuc_len)\n",
    "#         num_sgRNA_rows = int(max(sgRNA_rows) + 1)\n",
    "\n",
    "#     # Figure size and GridSpec: keep allele cells square via box aspect; keep equal widths via sharex\n",
    "#     cells_per_inch = 0.30  # width per base\n",
    "#     fig_w = plot_nuc_len * cells_per_inch\n",
    "\n",
    "#     ref_rows    = 1 + num_sgRNA_rows\n",
    "#     hm_rows     = N_ROWS\n",
    "#     twin_rows   = 1\n",
    "#     legend_rows = 0.8  # small strip for legend\n",
    "#     total_rows  = ref_rows + hm_rows + twin_rows + legend_rows\n",
    "\n",
    "#     # Height so that a unit of x (one base) equals a unit of y (one row) in ax_hm:\n",
    "#     # fig_h = fig_w * (total_rows / N_COLUMNS)\n",
    "#     fig_h = fig_w * (total_rows / float(N_COLUMNS))\n",
    "\n",
    "#     # Single GridSpec; no constrained_layout (avoids \"collapsed to zero\" warnings)\n",
    "#     fig = plt.figure(figsize=(fig_w, fig_h))\n",
    "#     gs = gridspec.GridSpec(\n",
    "#         4, N_COLUMNS,\n",
    "#         height_ratios=[ref_rows, hm_rows, twin_rows, legend_rows],\n",
    "#         figure=fig\n",
    "#     )\n",
    "\n",
    "#     # Axes with shared x to enforce identical widths and x-lims\n",
    "#     ax_hm      = fig.add_subplot(gs[1, :])\n",
    "#     ax_hm_ref  = fig.add_subplot(gs[0, :], sharex=ax_hm)\n",
    "#     ax_hm_twin = fig.add_subplot(gs[2, :], sharex=ax_hm)\n",
    "#     ax_lgd     = fig.add_subplot(gs[3, :])\n",
    "#     ax_lgd.axis('off')\n",
    "\n",
    "#     # Draw heatmaps (square=False here; we'll enforce aspect via set_box_aspect)\n",
    "#     custom_heatmap(ref_seq_hm,  annot=ref_seq_annot_hm,  annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm_ref,  vmin=0, vmax=5, square=False)\n",
    "#     custom_heatmap(X,           annot=np.array(annot),   annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm,     vmin=0, vmax=5, square=False, per_element_annot_kws=per_element_annot_kws)\n",
    "#     custom_heatmap(twin_seq_hm, annot=twin_seq_annot_hm, annot_kws={'size':16}, cmap=cmap, fmt='s',\n",
    "#                    ax=ax_hm_twin, vmin=0, vmax=5, square=False)\n",
    "\n",
    "#     # Make axes boxes respect data aspect (square cells) without shrinking width\n",
    "#     if hasattr(ax_hm, \"set_box_aspect\"):\n",
    "#         ax_hm.set_box_aspect(hm_rows / float(N_COLUMNS))\n",
    "#         ax_hm_ref.set_box_aspect(ref_rows / float(N_COLUMNS))\n",
    "#         ax_hm_twin.set_box_aspect(twin_rows / float(N_COLUMNS))\n",
    "#     else:\n",
    "#         # Fallback for older Matplotlib: only set equal on the main heatmap\n",
    "#         ax_hm.set_aspect(\"equal\", adjustable=\"box\")\n",
    "\n",
    "#     # Expand ref ylim to show sgRNA rows inside the ref axes\n",
    "#     if num_sgRNA_rows > 0:\n",
    "#         ax_hm_ref.set_ylim(-num_sgRNA_rows, 1)\n",
    "\n",
    "#     # Axes cosmetics\n",
    "#     ax_hm.yaxis.tick_right()\n",
    "#     ax_hm.yaxis.set_ticklabels(y_labels[::-1], rotation=True, va='center')\n",
    "#     ax_hm.xaxis.set_ticks([])\n",
    "\n",
    "#     ax_hm_ref.yaxis.tick_right()\n",
    "#     ax_hm_ref.xaxis.set_ticks([])\n",
    "#     ax_hm_ref.yaxis.set_ticklabels(['WT Reference'], rotation=True, va='center')\n",
    "\n",
    "#     ax_hm_twin.yaxis.tick_right()\n",
    "#     ax_hm_twin.xaxis.set_ticks([])\n",
    "#     ax_hm_twin.yaxis.set_ticklabels(['TwinPE Reference'], rotation=True, va='center')\n",
    "\n",
    "#     # Draw sgRNA overlays above the reference row\n",
    "#     if sgRNA_intervals and len(sgRNA_intervals) > 0:\n",
    "#         this_sgRNA_y_start  = -1 * num_sgRNA_rows\n",
    "#         this_sgRNA_y_height = num_sgRNA_rows - 0.3\n",
    "#         add_sgRNA_to_ax(ax_hm_ref, sgRNA_intervals,\n",
    "#                         sgRNA_y_start=this_sgRNA_y_start,\n",
    "#                         sgRNA_y_height=this_sgRNA_y_height,\n",
    "#                         amp_len=plot_nuc_len,\n",
    "#                         font_size='small',\n",
    "#                         clip_on=True,\n",
    "#                         sgRNA_names=sgRNA_names,\n",
    "#                         sgRNA_mismatches=sgRNA_mismatches,\n",
    "#                         x_offset=0,\n",
    "#                         label_at_zero=True,\n",
    "#                         sgRNA_rows=sgRNA_rows)\n",
    "\n",
    "#     # Outline insertion spans in red for each allele row\n",
    "#     for idx_row, lss in insertion_dict.items():\n",
    "#         for ls in lss:\n",
    "#             ax_hm.add_patch(\n",
    "#                 patches.Rectangle(\n",
    "#                     (ls[0], N_ROWS - idx_row - 1),\n",
    "#                     ls[1] - ls[0],\n",
    "#                     1,\n",
    "#                     linewidth=3,\n",
    "#                     edgecolor='r',\n",
    "#                     fill=False\n",
    "#                 )\n",
    "#             )\n",
    "\n",
    "#     # Draw predicted cleavage site(s) as vertical dashed line(s)\n",
    "#     if plot_cut_point:\n",
    "#         if cut_point_ind is None:\n",
    "#             cut_point_ind = [plot_nuc_len / 2]\n",
    "#         ax_hm.vlines(cut_point_ind, *ax_hm.get_ylim(), linestyles='dashed')\n",
    "\n",
    "#     # Styling context (unchanged)\n",
    "#     sns.set_context(rc={'axes.facecolor':'white','lines.markeredgewidth': 1,'mathtext.fontset' : 'stix','text.usetex':True,'text.latex.unicode':True} )\n",
    "\n",
    "#     # Legend proxies to communicate substitution/indel encodings\n",
    "#     proxies = [matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='black',\n",
    "#                     mec='none', marker=r'$\\mathbf{{{}}}$'.format('bold'), ms=18),\n",
    "#                matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "#                     mec='r', marker='s', ms=8, markeredgewidth=2.5),\n",
    "#                matplotlib.lines.Line2D([0], [0], linestyle='none', mfc='none',\n",
    "#                     mec='black', marker='_', ms=2,)]\n",
    "#     descriptions = ['Substitutions', 'Insertions', 'Deletions']\n",
    "\n",
    "#     if plot_cut_point:\n",
    "#         proxies.append(matplotlib.lines.Line2D([0], [1], linestyle='--', c='black', ms=6))\n",
    "#         descriptions.append('Predicted cleavage position')\n",
    "\n",
    "#     if category:\n",
    "#         category_disp = category.replace(\"_\", r\"\\ \")\n",
    "#         proxies.append(patches.Patch(color='none'))\n",
    "#         descriptions.append(rf\"$\\bf{{{category_disp}\\ Alleles}}$\")\n",
    "\n",
    "#     # Legend in its own row to avoid bbox_inches='tight' layout side-effects\n",
    "#     ax_lgd.legend(proxies, descriptions, numpoints=1, markerscale=2,\n",
    "#                   loc='center', ncol=1, fancybox=True, shadow=False)\n",
    "\n",
    "#     # Save/show without constrained/tight layout to keep axes widths identical\n",
    "#     if fig_filename_root is None:\n",
    "#         plt.show()\n",
    "#     else:\n",
    "#         fig.savefig(fig_filename_root + '.pdf')\n",
    "#         if SAVE_ALSO_PNG:\n",
    "#             fig.savefig(fig_filename_root + '.png')\n",
    "#     plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "affcb99f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_dataframe_allele_region(df_alleles, pegRNA_intervals, ref_seq, ref_aln_seq, twin_aln_seq, cut_points=None):\n",
    "#     \"\"\"\n",
    "#     Extract region of interest for plotting allele tables based on pegRNA positions on reference sequence.\n",
    "#     Returns modified reference sequence and cut points.\n",
    "#     \"\"\"\n",
    "#     if df_alleles.shape[0] == 0:\n",
    "#         return df_alleles\n",
    "#     if len(pegRNA_intervals) == 2:\n",
    "#         left_end = max(0, pegRNA_intervals[0][0] - 20)\n",
    "#         right_end = min(len(ref_seq), pegRNA_intervals[1][1] + 20)\n",
    "#         cut_points[0] = cut_points[0] - left_end\n",
    "#         cut_points[1] = cut_points[1] - left_end\n",
    "#     elif len(pegRNA_intervals) == 1:\n",
    "#         left_end = max(0, pegRNA_intervals[0][0] - 120)\n",
    "#         right_end = min(len(ref_seq), pegRNA_intervals[0][1] + 120)\n",
    "#         # cut_points = pegRNA_intervals[0][1] - 3\n",
    " \n",
    "    \n",
    "#     # Dont overwrite the original\n",
    "#     df = df_alleles.copy()\n",
    "\n",
    "#     # Slice the two sequence columns only\n",
    "#     df['Aligned_Sequence']   = df['Aligned_Sequence'].str.slice(left_end, right_end)\n",
    "#     df['Reference_Sequence'] = df['Reference_Sequence'].str.slice(left_end, right_end)\n",
    "\n",
    "#     # Aggregate counts by sequence + metadata\n",
    "#     df = (\n",
    "#         df.groupby(['Aligned_Sequence', 'Reference_Sequence', 'Read_Status',\n",
    "#                     'n_deleted', 'n_inserted', 'n_mutated'], dropna=False)\n",
    "#           .sum()\n",
    "#           .reset_index()\n",
    "#           .set_index('Aligned_Sequence')\n",
    "#     )\n",
    "\n",
    "#     df.sort_values(by=['#Reads', 'Aligned_Sequence', 'Reference_Sequence'],\n",
    "#                    ascending=[False, True, True],\n",
    "#                    inplace=True)\n",
    "\n",
    "#     df['Unedited'] = df['Read_Status'].eq('UNMODIFIED')\n",
    "\n",
    "#     return df, ref_seq[left_end:right_end], ref_aln_seq[left_end:right_end], twin_aln_seq[left_end:right_end], cut_points\n",
    "\n",
    "# ...existing code...\n",
    "def get_dataframe_allele_region(df_alleles, pegRNA_intervals, ref_seq, ref_aln_seq, twin_aln_seq, cut_points=None):\n",
    "    \"\"\"\n",
    "    Extract region of interest for plotting allele tables based on pegRNA positions on reference sequence.\n",
    "    Returns: (df_window, ref_seq_window, ref_aln_seq_window, twin_aln_seq_window, cut_points_window, pegRNA_intervals_window)\n",
    "    All returned coordinates (cut points, intervals) are shifted to the window (0-based) coordinates.\n",
    "    \"\"\"\n",
    "    if df_alleles.shape[0] == 0:\n",
    "        # Return empty frame and pass-through sequences, with None for coords\n",
    "        return df_alleles, ref_seq, ref_aln_seq, twin_aln_seq, cut_points, pegRNA_intervals\n",
    "\n",
    "    # Default padding\n",
    "    if len(pegRNA_intervals) == 2:\n",
    "        # left_end  = max(0, pegRNA_intervals[0][0] - 20)\n",
    "        left_end  = max(0, pegRNA_intervals[0][0] - 20)\n",
    "        right_end = min(len(ref_seq), pegRNA_intervals[1][1] + 20)\n",
    "    elif len(pegRNA_intervals) == 1:\n",
    "        # left_end  = max(0, pegRNA_intervals[0][0] - 120)\n",
    "        left_end  = 0\n",
    "        # right_end = min(len(ref_seq), pegRNA_intervals[0][1] + 120)\n",
    "        right_end = len(ref_seq)-1\n",
    "    else:\n",
    "        # Fallback: show full sequence\n",
    "        left_end, right_end = 0, len(ref_seq)\n",
    "\n",
    "    # Shift cut points into window coords (if provided)\n",
    "    cut_points_window = None\n",
    "    if cut_points is not None:\n",
    "        try:\n",
    "            cut_points_window = [cp - left_end for cp in cut_points]\n",
    "        except TypeError:\n",
    "            # Single integer\n",
    "            cut_points_window = [cut_points - left_end]\n",
    "\n",
    "    # Shift sgRNA intervals into window coords\n",
    "    pegRNA_intervals_window = []\n",
    "    if pegRNA_intervals is not None:\n",
    "        for (s, e) in pegRNA_intervals:\n",
    "            pegRNA_intervals_window.append((s - left_end, e - left_end))\n",
    "\n",
    "    # Dont overwrite the original\n",
    "    df = df_alleles.copy()\n",
    "\n",
    "    # Slice the two sequence columns only\n",
    "    df['Aligned_Sequence']   = df['Aligned_Sequence'].str.slice(left_end, right_end)\n",
    "    df['Reference_Sequence'] = df['Reference_Sequence'].str.slice(left_end, right_end)\n",
    "\n",
    "    # Aggregate counts by sequence + metadata\n",
    "    df = (\n",
    "        df.groupby(['Aligned_Sequence', 'Reference_Sequence', 'Read_Status',\n",
    "                    'n_deleted', 'n_inserted', 'n_mutated'], dropna=False)\n",
    "          .sum()\n",
    "          .reset_index()\n",
    "          .set_index('Aligned_Sequence')\n",
    "    )\n",
    "\n",
    "    df.sort_values(by=['#Reads', 'Aligned_Sequence', 'Reference_Sequence'],\n",
    "                   ascending=[False, True, True],\n",
    "                   inplace=True)\n",
    "\n",
    "    df['Unedited'] = df['Read_Status'].eq('UNMODIFIED')\n",
    "\n",
    "    return (\n",
    "        df,\n",
    "        ref_seq[left_end:right_end],\n",
    "        ref_aln_seq[left_end:right_end],\n",
    "        twin_aln_seq[left_end:right_end],\n",
    "        cut_points_window,\n",
    "        pegRNA_intervals_window\n",
    "    )\n",
    "# ...existing code..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e8310d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def setAlleleMatplotlibDefaults():\n",
    "    font = {'size': 22}\n",
    "    matplotlib.rc('font', **font)\n",
    "    matplotlib.use('AGG')\n",
    "    matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "    matplotlib.rcParams['ps.fonttype'] = 42\n",
    "    matplotlib.rcParams[\"font.sans-serif\"] = [\"Arial\", \"Liberation Sans\", \"Bitstream Vera Sans\"]\n",
    "    matplotlib.rcParams[\"font.family\"] = \"sans-serif\"\n",
    "    matplotlib.rcParams['axes.facecolor'] = 'white'\n",
    "    sns.set(style='white', font_scale=2.2)\n",
    "    plt.ioff()\n",
    "\n",
    "setAlleleMatplotlibDefaults()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "585320b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plotted category: WT\n",
      "Plotted category: Perfect_PE\n",
      "Plotted category: Right_Flap\n",
      "Plotted category: Imperfect_PE\n",
      "Plotted category: PE_Indel\n",
      "Plotted category: WT_Indel\n",
      "Plotted category: Left_Flap\n",
      "Plotted 7 plots\n"
     ]
    }
   ],
   "source": [
    "# Some alleles are longer or shorter than the reference sequence\n",
    "# These Allele tables plot smaller between full reference sequence and pegRNA +20bp padding reference\n",
    "# Can't extend window 20 bases left of first sgRNA because bc its beyond edge of reference sequence\n",
    "\n",
    "MIN_FREQUENCY = crispresso2_info['running_info']['args'].min_frequency_alleles_around_cut_to_plot\n",
    "MAX_N_ROWS = crispresso2_info['running_info']['args'].max_rows_alleles_around_cut_to_plot\n",
    "\n",
    "ref_name = crispresso2_info['results']['ref_names'][0]\n",
    "ref = crispresso2_info['results']['refs'][ref_name]\n",
    "# ref_seq = ref['sequence']\n",
    "\n",
    "b_extra = 3 # set based on number of bases cleaved in order to find spacerB in compound ref\n",
    "\n",
    "pegRNA_cut_points = ref['sgRNA_cut_points'].copy()\n",
    "# adjust spacerB cut point\n",
    "pegRNA_cut_points[1] = pegRNA_cut_points[1] - b_extra\n",
    "\n",
    "pegRNA_plot_cut_points = ref['sgRNA_plot_cut_points']\n",
    "pegRNA_intervals = ref['sgRNA_intervals']\n",
    "pegRNA_mismatches = ref['sgRNA_mismatches']\n",
    "pegRNA_names = ref['sgRNA_names']\n",
    "if len(ref['sgRNA_sequences']) == 2 and pegRNA_names == ['', '']:\n",
    "    pegRNA_names = ['pegRNA', 'pegRNA']\n",
    "\n",
    "plot_count = 0\n",
    "for cat in df_alleles['Category'].unique():\n",
    "    df_alleles_cat = df_alleles[df_alleles['Category'] == cat]\n",
    "\n",
    "    df_alleles_around_region, ref_seq_region, ref_aln_seq_region, twin_aln_seq_region, pegRNA_cut_points_region, pegRNA_intervals_region = get_dataframe_allele_region(\n",
    "        df_alleles_cat, \n",
    "        pegRNA_intervals, \n",
    "        wt_seq, \n",
    "        wt_aln_seq,\n",
    "        twin_aln_seq,\n",
    "        pegRNA_cut_points\n",
    "    )\n",
    "\n",
    "    X, annot, y_labels, insertion_dict, per_element_annot_kws, is_reference = prep_alleles_table(\n",
    "        df_alleles_around_region, \n",
    "        ref_seq_region, \n",
    "        ref_aln_seq_region, \n",
    "        twin_aln_seq_region,\n",
    "        MAX_N_ROWS, \n",
    "        MIN_FREQUENCY\n",
    "    )\n",
    "    \n",
    "    plot_alleles_heatmap(\n",
    "        reference_seq=ref_aln_seq_region,\n",
    "        X=X,\n",
    "        annot=annot,\n",
    "        y_labels=y_labels,\n",
    "        insertion_dict=insertion_dict,\n",
    "        per_element_annot_kws=per_element_annot_kws,\n",
    "        fig_filename_root=f\"figures/02_final_classify_reads_7cat/02_cat_{cat}\",\n",
    "        custom_colors=None,\n",
    "        SAVE_ALSO_PNG=False,\n",
    "        plot_cut_point=pegRNA_plot_cut_points,\n",
    "        cut_point_ind=pegRNA_cut_points_region,\n",
    "        sgRNA_intervals=pegRNA_intervals_region,\n",
    "        sgRNA_names=pegRNA_names,\n",
    "        sgRNA_mismatches=pegRNA_mismatches,\n",
    "        category=cat,\n",
    "        extend_left_non_gap=[0, b_extra]   # per-sgRNA non-gap extension\n",
    "\n",
    "    )\n",
    "\n",
    "# for cat in df_alleles['Category'].unique():\n",
    "#     df_alleles_cat = df_alleles[df_alleles['Category'] == cat]\n",
    "\n",
    "#     df_alleles_around_region, ref_seq_region, ref_aln_seq_region, twin_aln_seq_region, pegRNA_cut_points = get_dataframe_allele_region(\n",
    "#         df_alleles_cat, \n",
    "#         pegRNA_intervals, \n",
    "#         ref_seq, \n",
    "#         ref_aln_seq,\n",
    "#         twin_aln_seq,\n",
    "#         pegRNA_cut_points\n",
    "#     )\n",
    "\n",
    "#     X, annot, y_labels, insertion_dict, per_element_annot_kws, is_reference = prep_alleles_table(\n",
    "#         df_alleles_around_region, \n",
    "#         ref_seq_region, \n",
    "#         ref_aln_seq_region, \n",
    "#         twin_aln_seq_region,\n",
    "#         MAX_N_ROWS, \n",
    "#         MIN_FREQUENCY\n",
    "#     )\n",
    "\n",
    "#     plot_alleles_heatmap(\n",
    "#         reference_seq=ref_seq_region,\n",
    "#         X=X,\n",
    "#         annot=annot,\n",
    "#         y_labels=y_labels,\n",
    "#         insertion_dict=insertion_dict,\n",
    "#         per_element_annot_kws=per_element_annot_kws,\n",
    "#         fig_filename_root=f\"figures/anzalone_pe3_12bp_ins/01_cat_{cat}\",\n",
    "#         custom_colors=None,\n",
    "#         SAVE_ALSO_PNG=False,\n",
    "#         plot_cut_point=pegRNA_plot_cut_points,\n",
    "#         cut_point_ind=pegRNA_cut_points,\n",
    "#         sgRNA_intervals=pegRNA_intervals,\n",
    "#         sgRNA_names=pegRNA_names,\n",
    "#         sgRNA_mismatches=pegRNA_mismatches,\n",
    "#         category=cat\n",
    "#     )\n",
    "    plot_count += 1\n",
    "    print('Plotted category: ' + str(cat))\n",
    "\n",
    "print('Plotted ' + str(plot_count) + ' plots')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "696c8336",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "kc_crispresso",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
